---

title: Genodata module


keywords: fastai
sidebar: home_sidebar

summary: "read and extract genodata"
description: "read and extract genodata"
nb_path: "nbs/00_Genodata.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/00_Genodata.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_bgen" class="doc_header"><code>read_bgen</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L29" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_bgen</code>(<strong><code>file</code></strong>, <strong><code>sample_file</code></strong>=<em><code>None</code></em>, <strong><code>pybgen</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>the function to read genotype data</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_bim" class="doc_header"><code>read_bim</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L56" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_bim</code>(<strong><code>fn</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="bgen2dask" class="doc_header"><code>bgen2dask</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L63" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>bgen2dask</code>(<strong><code>bgen</code></strong>, <strong><code>index</code></strong>, <strong><code>step</code></strong>=<em><code>500</code></em>)</p>
</blockquote>
<p>The function to covert bgen to dask array</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="pybgen_region" class="doc_header"><code>pybgen_region</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L74" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>pybgen_region</code>(<strong><code>bgen</code></strong>, <strong><code>region</code></strong>, <strong><code>step</code></strong>=<em><code>100</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_bed" class="doc_header"><code>extract_bed</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L87" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_bed</code>(<strong><code>geno</code></strong>, <strong><code>idx</code></strong>, <strong><code>row</code></strong>=<em><code>True</code></em>, <strong><code>step</code></strong>=<em><code>500</code></em>, <strong><code>region</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<table>
<thead><tr>
<th></th>
<th>Type</th>
<th>Default</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>geno</code></strong></td>
<td></td>
<td></td>
<td><em>No Content</em></td>
</tr>
<tr>
<td><strong><code>idx</code></strong></td>
<td></td>
<td></td>
<td><em>No Content</em></td>
</tr>
<tr>
<td><strong><code>row</code></strong></td>
<td><code>bool</code></td>
<td><code>True</code></td>
<td><em>No Content</em></td>
</tr>
<tr>
<td><strong><code>step</code></strong></td>
<td><code>int</code></td>
<td><code>500</code></td>
<td><em>No Content</em></td>
</tr>
<tr>
<td><strong><code>region</code></strong></td>
<td><code>NoneType</code></td>
<td><code>None</code></td>
<td>row = True by variants, row = False by samples</td>
</tr>
</tbody>
</table>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="Genodata" class="doc_header"><code>class</code> <code>Genodata</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L108" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>Genodata</code>(<strong><code>geno_path</code></strong>, <strong><code>sample_path</code></strong>=<em><code>None</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="write_plink" class="doc_header"><code>write_plink</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L211" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>write_plink</code>(<strong><code>G</code></strong>, <strong><code>bed</code></strong>:<code>Union</code>[<code>str</code>, <code>Path</code>], <strong><code>bim</code></strong>:<code>Union</code>[<code>str</code>, <code>Path</code>, <code>NoneType</code>]=<em><code>None</code></em>, <strong><code>fam</code></strong>:<code>Union</code>[<code>str</code>, <code>Path</code>, <code>NoneType</code>]=<em><code>None</code></em>, <strong><code>row</code></strong>:<code>str</code>=<em><code>'variant'</code></em>, <strong><code>verbose</code></strong>:<code>bool</code>=<em><code>True</code></em>)</p>
</blockquote>
<p>Write PLINK 1 binary files into a data array.</p>
<p>A PLINK 1 binary file set consists of three files:</p>
<ul>
<li>BED: containing the genotype.</li>
<li>BIM: containing variant information.</li>
<li>FAM: containing sample information.</li>
</ul>
<p>The user must provide the genotype (dosage) via a :class:<code>xarray.DataArray</code> matrix
with data type :const:<code>numpy.float32</code> or :const:<code>numpy.float64</code>. That matrix must
have two named dimensions: <strong>sample</strong> and <strong>variant</strong>. The only allowed values for
the genotype are: :const:<code>0</code>, :const:<code>1</code>, :const:<code>2</code>, and :data:<code>math.nan</code>.</p>
<h2 id="Parameters">Parameters<a class="anchor-link" href="#Parameters"> </a></h2><p>G
    Genotype with bim, bed, and fam.
bed
    Path to a BED file.
bim
    Path to a BIM file.It defaults to :const:<code>None</code>, in which case it will try to be
    inferred.
fam
    Path to a FAM file. It defaults to :const:<code>None</code>, in which case it will try to
    be inferred.
major
    It can be either :const:<code>"sample"</code> or :const:<code>"variant"</code> (recommended and
    default). Specify the matrix layout on the BED file.
verbose
    :const:<code>True</code> for progress information; :const:<code>False</code> otherwise.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="write_fam" class="doc_header"><code>write_fam</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L278" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>write_fam</code>(<strong><code>filepath</code></strong>:<code>Path</code>, <strong><code>df</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="write_bim" class="doc_header"><code>write_bim</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L291" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>write_bim</code>(<strong><code>filepath</code></strong>:<code>Path</code>, <strong><code>df</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="write_bed" class="doc_header"><code>write_bed</code><a href="https://github.com/changebio/cugg/tree/master/cugg/genodata.py#L304" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>write_bed</code>(<strong><code>filepath</code></strong>:<code>Path</code>, <strong><code>G</code></strong>, <strong><code>row</code></strong>=<em><code>'variant'</code></em>, <strong><code>verbose</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>Write BED file.
It assumes that <code>X</code> is a variant-by-sample matrix.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">write_plink</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span><span class="s1">&#39;test.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>
Writing BED:   0%|          | 0/1 [00:00&lt;?, ?it/s]
Writing BED: 100%|██████████| 1/1 [00:04&lt;00:00,  4.77s/it]</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Writing FAM... </pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>done.
Writing BIM... done.
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">pandas_plink</span> <span class="kn">import</span> <span class="n">read_plink1_bin</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno1</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="s1">&#39;test.bed&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno1</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>bim:      chrom                  snp   cm        pos a0 a1     i
0         1    chr1:55039741:G:C  0.0   55039741  G  C     0
1         1    chr1:55039742:G:A  0.0   55039742  G  A     1
2         1    chr1:55039749:G:C  0.0   55039749  G  C     2
3         1    chr1:55039750:T:C  0.0   55039750  T  C     3
4         1    chr1:55039753:T:C  0.0   55039753  T  C     4
...     ...                  ...  ...        ... .. ..   ...
1408     11  chr11:116832956:T:C  0.0  116832956  T  C  1408
1409     11  chr11:116832976:C:G  0.0  116832976  C  G  1409
1410     11  chr11:116832977:T:G  0.0  116832977  T  G  1410
1411     11  chr11:116832978:T:G  0.0  116832978  T  G  1411
1412     11  chr11:116832980:T:C  0.0  116832980  T  C  1412

[1413 rows x 7 columns] 
 fam:            fid      iid father mother gender trait       i
0       1000019  1000019      0      0      2    -9       0
1       1000078  1000078      0      0      2    -9       1
2       1000081  1000081      0      0      1    -9       2
3       1000198  1000198      0      0      2    -9       3
4       1000210  1000210      0      0      1    -9       4
...         ...      ...    ...    ...    ...   ...     ...
168201  6025295  6025295      0      0      1    -9  168201
168202  6025319  6025319      0      0      2    -9  168202
168203  6025346  6025346      0      0      2    -9  168203
168204  6025363  6025363      0      0      1    -9  168204
168205  6025411  6025411      0      0      2    -9  168205

[168206 rows x 7 columns] 
 bed:dask.array&lt;transpose, shape=(1413, 168206), dtype=float32, chunksize=(1024, 1024), chunktype=numpy.ndarray&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Test">Test<a class="anchor-link" href="#Test"> </a></h1>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">geno_path</span> <span class="o">=</span><span class="s1">&#39;/home/dmc2245/UKBiobank/data/exome_files/project_VCF/072721_run/plink/ukb23156_c1.merged.filtered.bed&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">272741</span><span class="p">,</span><span class="mi">1213528</span><span class="o">-</span><span class="mi">900000</span><span class="p">]</span>
<span class="n">geno_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb23156_c5.merged.filtered.5_272741_1213528.bed&#39;</span>
<span class="n">sumstats_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/090321_UKBB_Hearing_aid_f3393_expandedwhite_6436cases_96601ctrl_PC1_2_f3393.regenie.snp_stats&#39;</span>
<span class="n">pheno_path</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">unr_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/UKB_genotypedatadownloaded083019.090221_sample_variant_qc_final_callrate90.filtered.extracted.white_europeans.filtered.092821_ldprun_unrelated.filtered.prune.txt&#39;</span>
<span class="n">imp_geno_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.bgen&#39;</span>
<span class="n">imp_sumstats_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/100521_UKBB_Hearing_aid_f3393_expandedwhite_15601cases_237318ctrl_500k_PC1_PC2_f3393.regenie.snp_stats&#39;</span>
<span class="n">imp_ref</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span>

<span class="n">output_sumstats</span> <span class="o">=</span> <span class="s1">&#39;test.snp_stats&#39;</span>
<span class="n">output_LD</span> <span class="o">=</span> <span class="s1">&#39;test_corr.csv&#39;</span>

<span class="c1">#main(region,geno_path,sumstats_path,pheno_path,unr_path,imp_geno_path,imp_sumstats_path,imp_ref,output_sumstats,output_LD)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">pandas_plink</span> <span class="kn">import</span> <span class="n">Chunk</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">Chunk</span><span class="p">(</span><span class="mi">512</span><span class="p">,</span><span class="mi">512</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>Chunk(nsamples=512, nvariants=512)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">exome_geno</span><span class="o">.</span><span class="n">extractbyvariants</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">snp</span><span class="p">[:</span><span class="mi">50</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">exome_geno</span><span class="o">.</span><span class="n">extractbysamples</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">fam</span><span class="o">.</span><span class="n">iid</span><span class="p">[:</span><span class="mi">60</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">cugg.sumstat</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">272741</span><span class="p">,</span> <span class="mi">1213528</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_sumstats</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="s1">&#39;/home/dmc2245/UKBiobank/results/REGENIE_results/results_imputed_data/2021_10_07_f3393_500K/100521_UKBB_Hearing_aid_f3393_expandedwhite_15601cases_237318ctrl_500k_PC1_PC2_f3393.regenie.snp_stats.gz&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_sumstats</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>sumstat:         CHR      POS REF ALT                SNP      BETA        SE         P
6767726    5   272851   A   G    chr5:272851:A:G  0.357496  0.888197  0.687318
6767727    5   272906   A   C    chr5:272906:A:C -0.003007  0.019764  0.879070
6767728    5   273143   A   G    chr5:273143:A:G -0.013693  0.016716  0.412684
6767729    5   273160   G   C    chr5:273160:G:C  0.235713  0.348772  0.499145
6767730    5   273534   C   T    chr5:273534:C:T  0.050095  0.139496  0.719509
...      ...      ...  ..  ..                ...       ...       ...       ...
6776191    5  1213094   C   T   chr5:1213094:C:T -0.015881  0.023298  0.495462
6776192    5  1213134   G   A   chr5:1213134:G:A -1.142280  1.344380  0.395509
6776193    5  1213223   C   T   chr5:1213223:C:T -0.003009  0.013631  0.825270
6776194    5  1213404   T  TC  chr5:1213404:T:TC -0.039146  0.117837  0.739735
6776195    5  1213510   C   T   chr5:1213510:C:T  0.009318  0.012922  0.470845

[8470 rows x 8 columns]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">bgen</span> <span class="o">=</span> <span class="n">PyBGEN</span><span class="p">(</span><span class="n">geno_file</span><span class="p">)</span>
<span class="n">sample_file</span> <span class="o">=</span> <span class="n">geno_file</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.bgen&#39;</span><span class="p">,</span> <span class="s1">&#39;.sample&#39;</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">sample_file</span><span class="p">):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">bgen_sample_path</span><span class="p">:</span><span class="n">r</span><span class="p">}):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cannot find the matching sample file ``</span><span class="si">{</span><span class="n">sample_file</span><span class="si">}</span><span class="s2">`` for ``</span><span class="si">{</span><span class="n">geno_file</span><span class="si">}</span><span class="s2">``.</span><span class="se">\n</span><span class="s2">You can specify path to sample file for all BGEN files using ``--bgen-sample-path``.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sample_file</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">bgen_sample_path</span><span class="p">:</span><span class="n">r</span><span class="p">}</span>
<span class="n">bgen_fam</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_file</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span><span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">bgen_fam</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;fid&#39;</span><span class="p">,</span><span class="s1">&#39;iid&#39;</span><span class="p">,</span><span class="s1">&#39;missing&#39;</span><span class="p">,</span><span class="s1">&#39;sex&#39;</span><span class="p">]</span>
<span class="n">geno</span> <span class="o">=</span> <span class="p">[</span><span class="n">bgen</span><span class="p">,</span><span class="n">bgen_fam</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">bgen_sample_path</span> <span class="o">=</span> <span class="s1">&#39;/home/dmc2245/UKBiobank_Yale_transfer/ukb39554_imputeddataset/ukb32285_imputedindiv.sample&#39;</span>
<span class="n">imput_geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="n">imp_geno_path</span><span class="p">,</span><span class="n">bgen_sample_path</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbyvariants</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">snp</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbysamples</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">fam</span><span class="o">.</span><span class="n">iid</span><span class="p">[</span><span class="mi">50</span><span class="p">:</span><span class="mi">100</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>bim:    chrom              snp   cm     pos a0 a1   i
10      5  chr5:273143:A:G  0.0  273143  A  G  10
11      5  chr5:273160:G:C  0.0  273160  G  C  11
12      5  chr5:273209:A:G  0.0  273209  A  G  12
13      5  chr5:273212:T:C  0.0  273212  T  C  13
14      5  chr5:273237:T:G  0.0  273237  T  G  14
15      5  chr5:273261:C:T  0.0  273261  C  T  15
16      5  chr5:273267:G:A  0.0  273267  G  A  16
17      5  chr5:273310:A:G  0.0  273310  A  G  17
18      5  chr5:273319:A:G  0.0  273319  A  G  18
19      5  chr5:273326:T:C  0.0  273326  T  C  19 
 fam:        fid      iid  missing  sex
50  1426232  1426232        0    1
51  3769979  3769979        0    1
52  1751431  1751431        0    2
53  4658109  4658109        0    2
54  3538677  3538677        0    2
55  3542585  3542585        0    2
56  3249692  3249692        0    2
57  2299845  2299845        0    1
58  4136172  4136172        0    2
59  5406314  5406314        0    1
60  5987848  5987848        0    2
61  3872614  3872614        0    1
62  4818195  4818195        0    2
63  1935571  1935571        0    2
64  3585286  3585286        0    1
65  2989866  2989866        0    2
66  1474609  1474609        0    1
67  1275311  1275311        0    2
68  5471349  5471349        0    2
69  3699731  3699731        0    2
70  2314069  2314069        0    2
71  4343716  4343716        0    2
72  1245364  1245364        0    2
73  4039100  4039100        0    1
74  4652328  4652328        0    1
75  2184919  2184919        0    1
76  3340731  3340731        0    1
77  5722885  5722885        0    1
78  2149616  2149616        0    1
79  3090686  3090686        0    1
80  5925070  5925070        0    2
81  2282972  2282972        0    1
82  4720630  4720630        0    2
83  1902948  1902948        0    2
84  2017232  2017232        0    1
85  3184972  3184972        0    2
86  1135664  1135664        0    1
87  5988158  5988158        0    2
88  5088374  5088374        0    1
89  3148263  3148263        0    2
90  3548968  3548968        0    2
91  5184236  5184236        0    2
92  2963226  2963226        0    2
93  2321078  2321078        0    2
94  2867519  2867519        0    1
95  4419678  4419678        0    2
96  5830444  5830444        0    2
97  2861021  2861021        0    2
98  2375434  2375434        0    1
99  1896578  1896578        0    1 
 bed:dask.array&lt;getitem, shape=(10, 50), dtype=int8, chunksize=(10, 50), chunktype=numpy.ndarray&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[5, 272741, 313528]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">pybgen</span> <span class="kn">import</span> <span class="n">PyBGEN</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">bgen</span> <span class="o">=</span> <span class="n">PyBGEN</span><span class="p">(</span><span class="n">imp_geno_path</span><span class="p">,</span><span class="n">probs_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pybgen_region</span><span class="p">(</span><span class="n">bgen</span><span class="p">,</span><span class="n">region</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result"><table>
    <tr>
        <td>
            <table>
                <thead>
                    <tr>
                        <td> </td>
                        <th> Array </th>
                        <th> Chunk </th>
                    </tr>
                </thead>
                <tbody>
                    
                    <tr>
                        <th> Bytes </th>
                        <td> 789.75 MiB </td>
                        <td> 46.48 MiB </td>
                    </tr>
                    
                    <tr>
                        <th> Shape </th>
                        <td> (1699, 487409) </td>
                        <td> (100, 487409) </td>
                    </tr>
                    <tr>
                        <th> Count </th>
                        <td> 34 Tasks </td>
                        <td> 17 Chunks </td>
                    </tr>
                    <tr>
                    <th> Type </th>
                    <td> int8 </td>
                    <td> numpy.ndarray </td>
                    </tr>
                </tbody>
            </table>
        </td>
        <td>
        <svg width="170" height="75" style="stroke:rgb(0,0,0);stroke-width:1" >

  <!-- Horizontal lines -->
  <line x1="0" y1="0" x2="120" y2="0" style="stroke-width:2" />
  <line x1="0" y1="1" x2="120" y2="1" />
  <line x1="0" y1="2" x2="120" y2="2" />
  <line x1="0" y1="4" x2="120" y2="4" />
  <line x1="0" y1="5" x2="120" y2="5" />
  <line x1="0" y1="7" x2="120" y2="7" />
  <line x1="0" y1="8" x2="120" y2="8" />
  <line x1="0" y1="10" x2="120" y2="10" />
  <line x1="0" y1="11" x2="120" y2="11" />
  <line x1="0" y1="13" x2="120" y2="13" />
  <line x1="0" y1="14" x2="120" y2="14" />
  <line x1="0" y1="16" x2="120" y2="16" />
  <line x1="0" y1="17" x2="120" y2="17" />
  <line x1="0" y1="19" x2="120" y2="19" />
  <line x1="0" y1="20" x2="120" y2="20" />
  <line x1="0" y1="22" x2="120" y2="22" />
  <line x1="0" y1="23" x2="120" y2="23" />
  <line x1="0" y1="25" x2="120" y2="25" style="stroke-width:2" />

  <!-- Vertical lines -->
  <line x1="0" y1="0" x2="0" y2="25" style="stroke-width:2" />
  <line x1="120" y1="0" x2="120" y2="25" style="stroke-width:2" />

  <!-- Colored Rectangle -->
  <polygon points="0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485" style="fill:#ECB172A0;stroke-width:0"/>

  <!-- Text -->
  <text x="60.000000" y="45.412617" font-size="1.0rem" font-weight="100" text-anchor="middle" >487409</text>
  <text x="140.000000" y="12.706308" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,140.000000,12.706308)">1699</text>
</svg>
        </td>
    </tr>
</table></div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">for</span> <span class="n">t</span><span class="p">,</span><span class="n">g</span> <span class="ow">in</span> <span class="n">bgen</span><span class="o">.</span><span class="n">iter_variants_in_region</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">&lt;</span><span class="mi">10</span> <span class="k">else</span> <span class="nb">str</span><span class="p">(</span><span class="n">region</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="n">region</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">region</span><span class="p">[</span><span class="mi">2</span><span class="p">]):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span> <span class="o">=</span> <span class="n">bgen</span><span class="o">.</span><span class="n">iter_variants</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">genos</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">bgen</span><span class="o">.</span><span class="n">nb_variants</span><span class="p">),</span><span class="n">bgen</span><span class="p">):</span>
    <span class="n">geno</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">100000</span> <span class="o">==</span><span class="mi">0</span><span class="p">:</span>
        <span class="n">geno</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>0 (&lt;Variant rs537688122 chr05:272856_A/G&gt;, array([[1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       ...,
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.]]))
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span>    <span class="n">genos</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">step</span><span class="p">):</span>
        <span class="n">onecode_geno</span> <span class="o">=</span> <span class="n">bgen</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="nb">min</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="n">i</span><span class="o">+</span><span class="n">step</span><span class="p">)])</span>  <span class="c1">#samples x variants</span>
        <span class="n">geno</span> <span class="o">=</span> <span class="n">onecode_geno</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
        <span class="n">genos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">geno</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="mi">1002</span> <span class="o">%</span><span class="k">10000</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>1002</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>PyBGEN(487,409 samples; 6,070,641 variants)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="n">tmp</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>array([[1.        , 0.        , 0.        ],
       [0.83921569, 0.16078431, 0.        ],
       [0.96078431, 0.03921569, 0.        ],
       ...,
       [1.        , 0.        , 0.        ],
       [0.00392157, 0.95686275, 0.03921569],
       [0.85490196, 0.14509804, 0.        ]])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aa</span> <span class="o">=</span> <span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>0    455061
1     31952
2       396
dtype: int64</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">t</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bgen</span><span class="o">.</span><span class="n">iter_variant_info</span><span class="p">()):</span>
    <span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">chrom</span><span class="p">),</span><span class="n">t</span><span class="o">.</span><span class="n">name</span><span class="p">,</span><span class="mf">0.0</span><span class="p">,</span><span class="n">t</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span><span class="n">t</span><span class="o">.</span><span class="n">a1</span><span class="p">,</span><span class="n">t</span><span class="o">.</span><span class="n">a2</span><span class="p">,</span><span class="n">i</span><span class="p">])</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span><span class="s1">&#39;snp&#39;</span><span class="p">,</span><span class="s1">&#39;cm&#39;</span><span class="p">,</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span><span class="s1">&#39;a0&#39;</span><span class="p">,</span><span class="s1">&#39;a1&#39;</span><span class="p">,</span><span class="s1">&#39;i&#39;</span><span class="p">])</span>
<span class="n">tmp</span><span class="o">.</span><span class="n">snp</span> <span class="o">=</span> <span class="s1">&#39;chr&#39;</span><span class="o">+</span><span class="n">tmp</span><span class="p">[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span><span class="s1">&#39;a0&#39;</span><span class="p">,</span><span class="s1">&#39;a1&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tmp</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>snp</th>
      <th>cm</th>
      <th>pos</th>
      <th>a0</th>
      <th>a1</th>
      <th>i</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5</td>
      <td>rs546183826</td>
      <td>0.0</td>
      <td>10043</td>
      <td>T</td>
      <td>A</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>5</td>
      <td>rs564373080</td>
      <td>0.0</td>
      <td>10055</td>
      <td>T</td>
      <td>A</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5</td>
      <td>rs528775171</td>
      <td>0.0</td>
      <td>10056</td>
      <td>A</td>
      <td>C</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5</td>
      <td>rs547354230</td>
      <td>0.0</td>
      <td>10058</td>
      <td>C</td>
      <td>A</td>
      <td>3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>rs562245928</td>
      <td>0.0</td>
      <td>10059</td>
      <td>C</td>
      <td>A</td>
      <td>4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>6070636</th>
      <td>5</td>
      <td>rs552848768</td>
      <td>0.0</td>
      <td>180902794</td>
      <td>G</td>
      <td>A</td>
      <td>6070636</td>
    </tr>
    <tr>
      <th>6070637</th>
      <td>5</td>
      <td>rs572660833</td>
      <td>0.0</td>
      <td>180902887</td>
      <td>T</td>
      <td>A</td>
      <td>6070637</td>
    </tr>
    <tr>
      <th>6070638</th>
      <td>5</td>
      <td>rs544891279</td>
      <td>0.0</td>
      <td>180903491</td>
      <td>G</td>
      <td>T</td>
      <td>6070638</td>
    </tr>
    <tr>
      <th>6070639</th>
      <td>5</td>
      <td>rs558980847</td>
      <td>0.0</td>
      <td>180904360</td>
      <td>A</td>
      <td>T</td>
      <td>6070639</td>
    </tr>
    <tr>
      <th>6070640</th>
      <td>5</td>
      <td>rs575676143</td>
      <td>0.0</td>
      <td>180904689</td>
      <td>T</td>
      <td>C</td>
      <td>6070640</td>
    </tr>
  </tbody>
</table>
<p>6070641 rows × 7 columns</p>
</div></div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">list</span><span class="p">(</span><span class="n">bgen</span><span class="o">.</span><span class="n">iter_variant_info</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;Variant rs546183826 chr05:10043_T/A&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">idx</span> <span class="o">=</span> <span class="n">imput_geno</span><span class="o">.</span><span class="n">idx</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">idx</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="ow">is</span> <span class="nb">bool</span><span class="p">:</span>
    <span class="n">pd_idx</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd_idx</span><span class="p">[</span><span class="n">pd_idx</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>8470</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">idx</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[10535, 10541, 10542, 10556, 10567, 10569, 10570, 10573, 10574]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imp_geno_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.bgen&#39;</span>
<span class="n">bgen</span> <span class="o">=</span> <span class="n">open_bgen</span><span class="p">(</span><span class="n">imp_geno_path</span><span class="p">)</span>
<span class="n">bgen</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>reading -- time=0:00:00.00, thread 1 of 1, part 1 of 1
</pre>
</div>
</div>

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>array([[[1., 0., 0.]],

       [[1., 0., 0.]],

       [[1., 0., 0.]],

       ...,

       [[1., 0., 0.]],

       [[1., 0., 0.]],

       [[1., 0., 0.]]])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imp_geno_path</span> <span class="o">=</span> <span class="s1">&#39;/mnt/mfs/statgen/archive/UKBiobank_Yale_transfer/ukb39554_imputeddataset/ukb_imp_chr5_v3.bgen&#39;</span>
<span class="n">bgen</span> <span class="o">=</span> <span class="n">open_bgen</span><span class="p">(</span><span class="n">imp_geno_path</span><span class="p">)</span>
<span class="n">bgen</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">KeyError</span>                                  Traceback (most recent call last)
<span class="ansi-green-fg">/tmp/1975928.1.all.q/ipykernel_31014/3592512439.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">      1</span> imp_geno_path <span class="ansi-blue-fg">=</span> <span class="ansi-blue-fg">&#39;/mnt/mfs/statgen/archive/UKBiobank_Yale_transfer/ukb39554_imputeddataset/ukb_imp_chr5_v3.bgen&#39;</span>
<span class="ansi-green-intense-fg ansi-bold">      2</span> bgen <span class="ansi-blue-fg">=</span> open_bgen<span class="ansi-blue-fg">(</span>imp_geno_path<span class="ansi-blue-fg">)</span>
<span class="ansi-green-fg">----&gt; 3</span><span class="ansi-red-fg"> </span>bgen<span class="ansi-blue-fg">.</span>read<span class="ansi-blue-fg">(</span><span class="ansi-cyan-fg">1</span><span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_bgen2.py</span> in <span class="ansi-cyan-fg">read</span><span class="ansi-blue-fg">(self, index, dtype, order, max_combinations, return_probabilities, return_missings, return_ploidies, num_threads)</span>
<span class="ansi-green-intense-fg ansi-bold">    529</span> 
<span class="ansi-green-intense-fg ansi-bold">    530</span>         max_combinations = (
<span class="ansi-green-fg">--&gt; 531</span><span class="ansi-red-fg">             </span>max_combinations <span class="ansi-green-fg">if</span> max_combinations <span class="ansi-green-fg">is</span> <span class="ansi-green-fg">not</span> <span class="ansi-green-fg">None</span> <span class="ansi-green-fg">else</span> self<span class="ansi-blue-fg">.</span>max_combinations
<span class="ansi-green-intense-fg ansi-bold">    532</span>         )  # Can&#39;t use &#39;or&#39; because it treats 0 as False
<span class="ansi-green-intense-fg ansi-bold">    533</span> 

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_bgen2.py</span> in <span class="ansi-cyan-fg">max_combinations</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-intense-fg ansi-bold">    711</span> 
<span class="ansi-green-intense-fg ansi-bold">    712</span>         &#34;&#34;&#34;
<span class="ansi-green-fg">--&gt; 713</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> self<span class="ansi-blue-fg">.</span>_metadata2_memmaps<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;max_combinations&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">[</span><span class="ansi-cyan-fg">0</span><span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    714</span> 
<span class="ansi-green-intense-fg ansi-bold">    715</span>     <span class="ansi-blue-fg">@</span>property

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_multimemmap.py</span> in <span class="ansi-cyan-fg">__getitem__</span><span class="ansi-blue-fg">(self, name)</span>
<span class="ansi-green-intense-fg ansi-bold">    200</span> 
<span class="ansi-green-intense-fg ansi-bold">    201</span>     <span class="ansi-green-fg">def</span> __getitem__<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span> name<span class="ansi-blue-fg">:</span> str<span class="ansi-blue-fg">)</span> <span class="ansi-blue-fg">-&gt;</span> np<span class="ansi-blue-fg">.</span>memmap<span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 202</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> self<span class="ansi-blue-fg">.</span>_name_to_memmap<span class="ansi-blue-fg">[</span>name<span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    203</span> 
<span class="ansi-green-intense-fg ansi-bold">    204</span>     def append_empty(

<span class="ansi-red-fg">KeyError</span>: &#39;max_combinations&#39;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bed</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;bgen_reader._bgen2.open_bgen at 0x2b1a4e207e80&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">bgen</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&lt;bgen_reader._bgen2.open_bgen at 0x2b1dfde98b20&gt;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">imput_geno</span><span class="o">.</span><span class="n">geno_in_stat</span><span class="p">(</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[False, False, False, False, False, False, False, False, False, False]
8470 [10529, 10535, 10541, 10542, 10556, 10567, 10569, 10570, 10573, 10574]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">KeyError</span>                                  Traceback (most recent call last)
<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/3359055265.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-fg">----&gt; 1</span><span class="ansi-red-fg"> </span>imput_geno<span class="ansi-blue-fg">.</span>geno_in_stat<span class="ansi-blue-fg">(</span>imput_sumstats<span class="ansi-blue-fg">.</span>ss<span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/2365980871.py</span> in <span class="ansi-cyan-fg">geno_in_stat</span><span class="ansi-blue-fg">(self, stat, notin)</span>
<span class="ansi-green-intense-fg ansi-bold">     21</span>         <span class="ansi-blue-fg">&#39;&#39;&#39;The function to find an overlap region between geno data with sumstat&#39;&#39;&#39;</span>
<span class="ansi-green-intense-fg ansi-bold">     22</span>         variants <span class="ansi-blue-fg">=</span> stat<span class="ansi-blue-fg">.</span>SNP
<span class="ansi-green-fg">---&gt; 23</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>extractbyvariants<span class="ansi-blue-fg">(</span>variants<span class="ansi-blue-fg">,</span>notin<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     24</span> 
<span class="ansi-green-intense-fg ansi-bold">     25</span> 

<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/2365980871.py</span> in <span class="ansi-cyan-fg">extractbyvariants</span><span class="ansi-blue-fg">(self, variants, notin)</span>
<span class="ansi-green-intense-fg ansi-bold">     44</span>             <span class="ansi-green-fg">raise</span> ValueError<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#39;The extraction is empty&#39;</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     45</span>         <span class="ansi-red-fg">#update bim,bed</span>
<span class="ansi-green-fg">---&gt; 46</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>extractbyidx<span class="ansi-blue-fg">(</span>idx<span class="ansi-blue-fg">,</span>row<span class="ansi-blue-fg">=</span><span class="ansi-green-fg">True</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     47</span> 
<span class="ansi-green-intense-fg ansi-bold">     48</span>     <span class="ansi-green-fg">def</span> extractbysamples<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span>samples<span class="ansi-blue-fg">,</span>notin<span class="ansi-blue-fg">=</span><span class="ansi-green-fg">False</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span> <span class="ansi-red-fg">#samples is list or pd.Series</span>

<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/2365980871.py</span> in <span class="ansi-cyan-fg">extractbyidx</span><span class="ansi-blue-fg">(self, idx, row)</span>
<span class="ansi-green-intense-fg ansi-bold">     74</span>             <span class="ansi-green-fg">else</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">     75</span>                 self<span class="ansi-blue-fg">.</span>fam <span class="ansi-blue-fg">=</span> self<span class="ansi-blue-fg">.</span>fam<span class="ansi-blue-fg">.</span>iloc<span class="ansi-blue-fg">[</span>idx<span class="ansi-blue-fg">]</span>
<span class="ansi-green-fg">---&gt; 76</span><span class="ansi-red-fg">         </span>self<span class="ansi-blue-fg">.</span>bed <span class="ansi-blue-fg">=</span> extract_bed<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">.</span>bed<span class="ansi-blue-fg">,</span>idx<span class="ansi-blue-fg">,</span>row<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     77</span> 

<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/2305036667.py</span> in <span class="ansi-cyan-fg">extract_bed</span><span class="ansi-blue-fg">(geno, idx, row, step)</span>
<span class="ansi-green-intense-fg ansi-bold">     13</span>                 pd_idx <span class="ansi-blue-fg">=</span> pd<span class="ansi-blue-fg">.</span>Series<span class="ansi-blue-fg">(</span>idx<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     14</span>                 idx <span class="ansi-blue-fg">=</span> list<span class="ansi-blue-fg">(</span>pd_idx<span class="ansi-blue-fg">[</span>pd_idx<span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">.</span>index<span class="ansi-blue-fg">)</span>
<span class="ansi-green-fg">---&gt; 15</span><span class="ansi-red-fg">             </span>geno <span class="ansi-blue-fg">=</span> bgen2dask<span class="ansi-blue-fg">(</span>geno<span class="ansi-blue-fg">,</span>idx<span class="ansi-blue-fg">,</span>step<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     16</span>         <span class="ansi-green-fg">else</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-intense-fg ansi-bold">     17</span>             geno <span class="ansi-blue-fg">=</span> geno<span class="ansi-blue-fg">.</span>read<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">)</span> <span class="ansi-red-fg"># read all variants</span>

<span class="ansi-green-fg">/tmp/1967998.1.high_mem.q/ipykernel_3740/738061970.py</span> in <span class="ansi-cyan-fg">bgen2dask</span><span class="ansi-blue-fg">(bgen, index, step)</span>
<span class="ansi-green-intense-fg ansi-bold">      6</span>     print<span class="ansi-blue-fg">(</span>n<span class="ansi-blue-fg">,</span>index<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">:</span><span class="ansi-cyan-fg">10</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">      7</span>     <span class="ansi-green-fg">for</span> i <span class="ansi-green-fg">in</span> range<span class="ansi-blue-fg">(</span><span class="ansi-cyan-fg">0</span><span class="ansi-blue-fg">,</span>n<span class="ansi-blue-fg">,</span>step<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">----&gt; 8</span><span class="ansi-red-fg">         </span>onecode_geno <span class="ansi-blue-fg">=</span> bgen<span class="ansi-blue-fg">.</span>read<span class="ansi-blue-fg">(</span>index<span class="ansi-blue-fg">[</span>i<span class="ansi-blue-fg">:</span>min<span class="ansi-blue-fg">(</span>n<span class="ansi-blue-fg">,</span>i<span class="ansi-blue-fg">+</span>step<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">)</span>  <span class="ansi-red-fg">#samples x variants</span>
<span class="ansi-green-intense-fg ansi-bold">      9</span>         geno <span class="ansi-blue-fg">=</span> onecode_geno<span class="ansi-blue-fg">.</span>argmax<span class="ansi-blue-fg">(</span>axis<span class="ansi-blue-fg">=</span><span class="ansi-cyan-fg">2</span><span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">.</span>astype<span class="ansi-blue-fg">(</span>np<span class="ansi-blue-fg">.</span>int8<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     10</span>         genos<span class="ansi-blue-fg">.</span>append<span class="ansi-blue-fg">(</span>da<span class="ansi-blue-fg">.</span>from_array<span class="ansi-blue-fg">(</span>geno<span class="ansi-blue-fg">)</span><span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_bgen2.py</span> in <span class="ansi-cyan-fg">read</span><span class="ansi-blue-fg">(self, index, dtype, order, max_combinations, return_probabilities, return_missings, return_ploidies, num_threads)</span>
<span class="ansi-green-intense-fg ansi-bold">    529</span> 
<span class="ansi-green-intense-fg ansi-bold">    530</span>         max_combinations = (
<span class="ansi-green-fg">--&gt; 531</span><span class="ansi-red-fg">             </span>max_combinations <span class="ansi-green-fg">if</span> max_combinations <span class="ansi-green-fg">is</span> <span class="ansi-green-fg">not</span> <span class="ansi-green-fg">None</span> <span class="ansi-green-fg">else</span> self<span class="ansi-blue-fg">.</span>max_combinations
<span class="ansi-green-intense-fg ansi-bold">    532</span>         )  # Can&#39;t use &#39;or&#39; because it treats 0 as False
<span class="ansi-green-intense-fg ansi-bold">    533</span> 

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_bgen2.py</span> in <span class="ansi-cyan-fg">max_combinations</span><span class="ansi-blue-fg">(self)</span>
<span class="ansi-green-intense-fg ansi-bold">    711</span> 
<span class="ansi-green-intense-fg ansi-bold">    712</span>         &#34;&#34;&#34;
<span class="ansi-green-fg">--&gt; 713</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> self<span class="ansi-blue-fg">.</span>_metadata2_memmaps<span class="ansi-blue-fg">[</span><span class="ansi-blue-fg">&#34;max_combinations&#34;</span><span class="ansi-blue-fg">]</span><span class="ansi-blue-fg">[</span><span class="ansi-cyan-fg">0</span><span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    714</span> 
<span class="ansi-green-intense-fg ansi-bold">    715</span>     <span class="ansi-blue-fg">@</span>property

<span class="ansi-green-fg">~/miniconda3/lib/python3.8/site-packages/bgen_reader/_multimemmap.py</span> in <span class="ansi-cyan-fg">__getitem__</span><span class="ansi-blue-fg">(self, name)</span>
<span class="ansi-green-intense-fg ansi-bold">    200</span> 
<span class="ansi-green-intense-fg ansi-bold">    201</span>     <span class="ansi-green-fg">def</span> __getitem__<span class="ansi-blue-fg">(</span>self<span class="ansi-blue-fg">,</span> name<span class="ansi-blue-fg">:</span> str<span class="ansi-blue-fg">)</span> <span class="ansi-blue-fg">-&gt;</span> np<span class="ansi-blue-fg">.</span>memmap<span class="ansi-blue-fg">:</span>
<span class="ansi-green-fg">--&gt; 202</span><span class="ansi-red-fg">         </span><span class="ansi-green-fg">return</span> self<span class="ansi-blue-fg">.</span>_name_to_memmap<span class="ansi-blue-fg">[</span>name<span class="ansi-blue-fg">]</span>
<span class="ansi-green-intense-fg ansi-bold">    203</span> 
<span class="ansi-green-intense-fg ansi-bold">    204</span>     def append_empty(

<span class="ansi-red-fg">KeyError</span>: &#39;max_combinations&#39;</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">read_bgen</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/archive/UKBiobank_Yale_transfer/ukb39554_imputeddataset/ukb_imp_chr5_v3.bgen&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(         chrom                 snp        pos a0 a1
 0            5      chr5:10043:T:A      10043  T  A
 1            5      chr5:10055:T:A      10055  T  A
 2            5      chr5:10056:A:C      10056  A  C
 3            5      chr5:10058:C:A      10058  C  A
 4            5      chr5:10059:C:A      10059  C  A
 ...        ...                 ...        ... .. ..
 6070636      5  chr5:180902794:G:A  180902794  G  A
 6070637      5  chr5:180902887:T:A  180902887  T  A
 6070638      5  chr5:180903491:G:T  180903491  G  T
 6070639      5  chr5:180904360:A:T  180904360  A  T
 6070640      5  chr5:180904689:T:C  180904689  T  C
 
 [6070641 rows x 5 columns],
 None,
 &lt;bgen_reader._bgen2.open_bgen at 0x2b1a4e295c70&gt;)</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

