---

title: Scalable pipeline for computing LD matrix in big sample phenotype


keywords: fastai
sidebar: home_sidebar



nb_path: "nbs/index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="4-modules">4 modules<a class="anchor-link" href="#4-modules"> </a></h3><ul>
<li>Genodata</li>
<li>Sumstats</li>
<li>Liftover</li>
<li>LDmatrix</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Install">Install<a class="anchor-link" href="#Install"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><code>pip install cugg</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="How-to-use">How to use<a class="anchor-link" href="#How-to-use"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">vcf</span> <span class="o">=</span><span class="s1">&#39;/home/yh3455/Github/SEQLinkage/MWE/small_sample_ii_coding.vcf.gz&#39;</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">vcf_liftover</span><span class="p">(</span><span class="n">vcf</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">!</span>which python
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>/home/yh3455/miniconda3/bin/python
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">region</span> <span class="o">=</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">272741</span><span class="p">,</span><span class="mi">1213528</span><span class="o">-</span><span class="mi">900000</span><span class="p">]</span>
<span class="n">geno_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb23156_c5.merged.filtered.5_272741_1213528.bed&#39;</span>
<span class="n">sumstats_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/090321_UKBB_Hearing_aid_f3393_expandedwhite_6436cases_96601ctrl_PC1_2_f3393.regenie.snp_stats&#39;</span>
<span class="n">pheno_path</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">unr_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/UKB_genotypedatadownloaded083019.090221_sample_variant_qc_final_callrate90.filtered.extracted.white_europeans.filtered.092821_ldprun_unrelated.filtered.prune.txt&#39;</span>
<span class="n">imp_geno_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.bgen&#39;</span>
<span class="n">imp_sumstats_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/100521_UKBB_Hearing_aid_f3393_expandedwhite_15601cases_237318ctrl_500k_PC1_PC2_f3393.regenie.snp_stats.gz&#39;</span>
<span class="n">imp_ref</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span>
<span class="n">bgen_sample_path</span> <span class="o">=</span> <span class="s1">&#39;MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.sample&#39;</span>
<span class="n">output_sumstats</span> <span class="o">=</span> <span class="s1">&#39;test.snp_stats.gz&#39;</span>
<span class="n">output_LD</span> <span class="o">=</span> <span class="s1">&#39;test_corr.csv.gz&#39;</span>

<span class="c1">#main(region,geno_path,sumstats_path,pheno_path,unr_path,imp_geno_path,imp_sumstats_path,imp_ref,output_sumstats,output_LD)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span>    <span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">region</span><span class="p">,</span><span class="n">geno_path</span><span class="p">,</span><span class="n">sumstats_path</span><span class="p">,</span><span class="n">pheno_path</span><span class="p">,</span><span class="n">unr_path</span><span class="p">,</span><span class="n">imp_geno_path</span><span class="p">,</span><span class="n">imp_sumstats_path</span><span class="p">,</span><span class="n">imp_ref</span><span class="p">,</span><span class="n">output_sumstats</span><span class="p">,</span><span class="n">output_LD</span><span class="p">,</span><span class="n">bgen_sample_path</span><span class="p">):</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1. Preprocess sumstats (regenie format) and extract it from a region&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pheno_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Load phenotype file</span>
            <span class="n">pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">unr_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Load unrelated sample file</span>
            <span class="n">unr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">unr_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">quotechar</span><span class="o">=</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span>  
        <span class="c1"># Load the file of summary statistics and standardize it.</span>
        <span class="n">exome_sumstats</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">sumstats_path</span><span class="p">)</span>
        <span class="n">exome_geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="n">geno_path</span><span class="p">,</span><span class="n">bgen_sample_path</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.1. Region extraction&#39;</span><span class="p">)</span>
        <span class="n">exome_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
        <span class="n">exome_geno</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
        <span class="n">exome_sumstats</span><span class="o">.</span><span class="n">match_ss</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">bim</span><span class="p">)</span>
        <span class="n">exome_geno</span><span class="o">.</span><span class="n">geno_in_stat</span><span class="p">(</span><span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">imp_geno_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1">#two genotype data</span>
            <span class="n">imput_sumstats</span> <span class="o">=</span> <span class="n">Sumstat</span><span class="p">(</span><span class="n">imp_sumstats_path</span><span class="p">)</span>
            <span class="n">imput_geno</span> <span class="o">=</span> <span class="n">Genodata</span><span class="p">(</span><span class="n">imp_geno_path</span><span class="p">,</span><span class="n">bgen_sample_path</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">imp_ref</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">match_ss</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="p">)</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">geno_in_stat</span><span class="p">(</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.2. LiftOver the region&#39;</span><span class="p">)</span>
                <span class="n">hg38toimpref</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span><span class="n">imp_ref</span><span class="p">)</span>
                <span class="n">imp_region</span> <span class="o">=</span> <span class="n">hg38toimpref</span><span class="o">.</span><span class="n">region_liftover</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">imp_region</span><span class="p">)</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbyregion</span><span class="p">(</span><span class="n">imp_region</span><span class="p">)</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">match_ss</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="p">)</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">geno_in_stat</span><span class="p">(</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>

                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.3. Regional SNPs Liftover&#39;</span><span class="p">)</span>
                <span class="n">impreftohg38</span> <span class="o">=</span> <span class="n">Liftover</span><span class="p">(</span><span class="n">imp_ref</span><span class="p">,</span><span class="s1">&#39;hg38&#39;</span><span class="p">)</span> <span class="c1">#oppsite with hg38toimpref</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span> <span class="o">=</span> <span class="n">impreftohg38</span><span class="o">.</span><span class="n">bim_liftover</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="p">)</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="o">.</span><span class="n">POS</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">pos</span><span class="p">)</span>
                <span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="o">.</span><span class="n">SNP</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">snp</span><span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1.1.1 Get exome unique sumstats and geno and Combine sumstats&#39;</span><span class="p">)</span>
            <span class="n">snp_match</span> <span class="o">=</span> <span class="n">compare_snps</span><span class="p">(</span><span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">,</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
            <span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span> <span class="o">=</span> <span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">snp_match</span><span class="o">.</span><span class="n">qidx</span><span class="p">[</span><span class="n">snp_match</span><span class="o">.</span><span class="n">exact</span><span class="o">==</span><span class="kc">False</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()]</span> <span class="c1">#remove by exact match. can be improved.</span>
            <span class="n">exome_sumstats</span><span class="o">.</span><span class="n">extractbyvariants</span><span class="p">(</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span><span class="n">notin</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">exome_geno</span><span class="o">.</span><span class="n">geno_in_stat</span><span class="p">(</span><span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">)</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">exome_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">,</span><span class="n">imput_sumstats</span><span class="o">.</span><span class="n">ss</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#one genotype data</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="n">exome_sumstats</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;2. Remove relative samples&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">unr_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">exome_geno</span><span class="o">.</span><span class="n">geno_in_unr</span><span class="p">(</span><span class="n">unr</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">imp_geno_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">imput_geno</span><span class="o">.</span><span class="n">geno_in_unr</span><span class="p">(</span><span class="n">unr</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Warning:There is no file of relative sample. All sample are included in computing LD matrix&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">pheno_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Warning: This function has been implementd yet.&#39;</span><span class="p">)</span>
            <span class="k">pass</span> <span class="c1">#sld and pld</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;3. Calculate LD matrix&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">imp_geno_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">cor_da</span> <span class="o">=</span> <span class="n">geno_corr</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">xx</span> <span class="o">=</span> <span class="n">geno_corr</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
            <span class="n">yy</span> <span class="o">=</span> <span class="n">geno_corr</span><span class="p">(</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">step</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>

            <span class="n">imput_fam</span> <span class="o">=</span> <span class="n">imput_geno</span><span class="o">.</span><span class="n">fam</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">imput_fam</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">imput_fam</span><span class="o">.</span><span class="n">iid</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
            <span class="n">imput_fam</span><span class="p">[</span><span class="s1">&#39;i&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">imput_fam</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
            <span class="n">imput_fam_comm</span> <span class="o">=</span> <span class="n">imput_fam</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">fam</span><span class="o">.</span><span class="n">iid</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))]</span>
            <span class="n">imput_geno</span><span class="o">.</span><span class="n">extractbyidx</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">imput_fam_comm</span><span class="o">.</span><span class="n">i</span><span class="p">),</span><span class="n">row</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">xy</span> <span class="o">=</span> <span class="n">geno_corr</span><span class="p">(</span><span class="n">exome_geno</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">imput_geno</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">step</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
            <span class="n">cor_da</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">da</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">xx</span><span class="p">,</span><span class="n">xy</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span><span class="n">da</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">xy</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">yy</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;4. Output sumstats and LD matrix&#39;</span><span class="p">)</span>
        <span class="n">index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">sumstats</span><span class="o">.</span><span class="n">SNP</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">shorten_id</span><span class="p">))</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">SNP</span> <span class="o">=</span> <span class="n">index</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">sumstats</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_sumstats</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>

        <span class="n">corr</span> <span class="o">=</span> <span class="n">cor_da</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
        <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">corr</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">corr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">corr</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">index</span><span class="p">)</span>
        <span class="n">corr</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_LD</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">main</span><span class="p">(</span><span class="n">region</span><span class="p">,</span><span class="n">geno_path</span><span class="p">,</span><span class="n">sumstats_path</span><span class="p">,</span><span class="n">pheno_path</span><span class="p">,</span><span class="n">unr_path</span><span class="p">,</span><span class="n">imp_geno_path</span><span class="p">,</span><span class="n">imp_sumstats_path</span><span class="p">,</span><span class="n">imp_ref</span><span class="p">,</span><span class="n">output_sumstats</span><span class="p">,</span><span class="n">output_LD</span><span class="p">,</span><span class="n">bgen_sample_path</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>1. Preprocess sumstats (regenie format) and extract it from a region
1.1. Region extraction
this region [5, 272741, 313528] has 498 SNPs
Total SNPs 119 . Flip SNPs 118
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/home/yh3455/miniconda3/lib/python3.8/site-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>1.2. LiftOver the region
this region (5, 272856, 313643) has 1736 SNPs
Total SNPs 385 . Flip SNPs 0
1.3. Regional SNPs Liftover
1.1.1 Get exome unique sumstats and geno and Combine sumstats
keep   exact  flip   reverse  both   complement
False  False  False  False    False  False         106
True   False  True   False    False  False          12
       True   False  False    False  False           1
dtype: int64
2. Remove relative samples
3. Calculate LD matrix
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>/mnt/mfs/statgen/yin/Github/LDtools/LDtools/ldmatrix.py:29: RuntimeWarning: invalid value encountered in true_divide
  geno_i = (geno_i - np.nanmean(geno_i,axis=0)[None,:])/np.nanstd(geno_i,axis=0)[None,:]
/mnt/mfs/statgen/yin/Github/LDtools/LDtools/genodata.py:74: PerformanceWarning: Slicing with an out-of-order index is generating 22098 times more chunks
  geno = geno[:,idx]
/mnt/mfs/statgen/yin/Github/LDtools/LDtools/ldmatrix.py:67: RuntimeWarning: invalid value encountered in true_divide
  geno_i = (geno_i - np.nanmean(geno_i,axis=0)[None,:])/np.nanstd(geno_i,axis=0)[None,:]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>4. Output sumstats and LD matrix
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

