{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a661dd1-690b-4b58-bbf1-be85586ae133",
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "408c13de-4d70-4b1b-b96d-c5710fe5b2b6",
   "metadata": {},
   "source": [
    "```pip install biopython```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16d5217f-7ecd-42cd-a12a-ea2f3f1ea8ed",
   "metadata": {},
   "source": [
    "## Two scenarios\n",
    "1. one dataset (vcf) be used in different softwares (regenie and fine mapping). we only need to consider flip and all snps in sumstat must be in genodata.\n",
    "2. two datasets. They may have strand flips and reverse reference alleles, or both. So we need to consider all kinds of issues. I think the best way is to show all the potential results\n",
    "    (multi SNPs in one postion and SNPs more than one base) and print a summary result, which we may get some very important info in there.\n",
    "    \n",
    "For region extraction case, I think we only need to consider scenario 1, which make genotype consistent in different softwares by shifting the sign of beta in sumstats.\n",
    "For merging exome and imput data, it is not necessary to consider the SNPs' overlapping in defferent datasets. If two SNPs are the same, both of them will be show in a credit set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "ba5461c6-ff76-45bb-b4ac-5b049b54a8cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "bbf363e0-04d0-4774-a011-20984b63269f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "import dask.array as da\n",
    "from cugg.liftover import *\n",
    "from cugg.genodata import *\n",
    "from cugg.sumstat import *\n",
    "from cugg.ldmatrix import *\n",
    "from cugg.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "cf0f2a11-b414-4a54-90fe-0ad6e6ea3d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.Seq import Seq\n",
    "import warnings\n",
    "import yaml\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9be10438-e312-46ec-9a9a-4f5618c4318d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def check_indels(query):\n",
    "    #check duplicated indels and remove them.\n",
    "    indels = query.index.duplicated(keep=False)\n",
    "    if(indels.any()):\n",
    "        warnings.warn(\"There are SNPs {}: REF:ALT = ALT:REF. They will be removed.\".format(sum(indels)))\n",
    "        query = query[~indels]\n",
    "    return query\n",
    "\n",
    "def load_yaml(yaml_file):\n",
    "    with open(yaml_file, \"r\") as stream:\n",
    "        try:\n",
    "            yml = yaml.safe_load(stream)\n",
    "        except yaml.YAMLError as exc:\n",
    "            print(exc)\n",
    "    return yml\n",
    "\n",
    "def parse_input(yml_input):\n",
    "    input_dict = {}\n",
    "    for i in yml_input:\n",
    "        for name in glob.glob(list(i.keys())[0]):\n",
    "            input_dict[name] = list(i.values())[0].copy()\n",
    "    return input_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "411f2798-5d0e-486e-a7fc-fc7de27c4bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def match_ss_with_bim(query,subject):\n",
    "    '''snp match case in one dataset'''\n",
    "    query = query.itertuples()\n",
    "    subject = subject.itertuples()\n",
    "    qi,si = next(query,None),next(subject,None)\n",
    "    flip=[]\n",
    "    while(qi and si):\n",
    "        if qi.CHR>si.chrom:\n",
    "            si = next(subject,None)\n",
    "            continue\n",
    "        elif qi.CHR<si.chrom:\n",
    "            raise Exception(qi,'is not in genodata')\n",
    "        else:\n",
    "            if qi.POS>si.pos:\n",
    "                si = next(subject,None)\n",
    "                continue\n",
    "            elif qi.POS<si.pos:\n",
    "                raise Exception(qi,'is not in genodata')\n",
    "            else:\n",
    "                #Fix me\n",
    "                #same pos has multiple snps\n",
    "                #query compare with each of them in subject\n",
    "                if qi.ALT == si.a1 and qi.REF == si.a0:\n",
    "                    flip.append(False)\n",
    "                elif qi.ALT == si.a0 and qi.REF == si.a1:\n",
    "                    flip.append(True)\n",
    "                else:\n",
    "                    si = next(subject,None)\n",
    "                    continue\n",
    "                qi = next(query,None)\n",
    "                si = next(subject,None)\n",
    "    return flip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "a03cf472-6d8e-4249-825d-5abf17649b95",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def check_ss(ss,bim):\n",
    "    ss.CHR = ss.CHR.astype(int)\n",
    "    bim.chrom = bim.chrom.astype(int)\n",
    "    flip = np.array(match_ss_with_bim(ss,bim))\n",
    "    print('Total SNPs',len(flip),'. Flip SNPs',sum(flip))\n",
    "    #shift ref and alt for fliped snps and change snp's name and the sign of beta\n",
    "    if flip.any():\n",
    "        fss = ss[flip]\n",
    "        ref = fss.ALT.copy()\n",
    "        alt = fss.REF.copy()\n",
    "        fss.REF = ref\n",
    "        fss.ALT =alt\n",
    "        fss.SNP = 'chr'+fss[['CHR','POS','REF','ALT']].astype(str).agg(':'.join, axis=1)\n",
    "        fss.BETA = -fss.BETA\n",
    "        ss = pd.concat([ss[~flip],fss]).sort_index()\n",
    "    return ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "3e86bc3a-7237-4d33-804c-98261e04f475",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def compare_snps(query,subject,only_match=True):\n",
    "    '''\n",
    "    input: query and subject, two data frame with the first five column: chr, pos, snp, a0, a1\n",
    "    output: data frame included six boolean columns (keep,exact,flip,reverse,both, complement) and two columns of the index query and subject.\n",
    "    '''\n",
    "    smry = []\n",
    "    query = query.itertuples()\n",
    "    subject = subject.itertuples()\n",
    "    pre_s = None\n",
    "    qi,si = next(query,None),next(subject,None)\n",
    "    multi_snps = []\n",
    "    while(qi and si):\n",
    "        if qi[1]>si[1]:\n",
    "            si = next(subject,None)\n",
    "        elif qi[1]<si[1]:\n",
    "            smry.append([False]*5+[qi[0],-1])\n",
    "            qi = next(query,None)\n",
    "        else:\n",
    "            if qi[2]>si[2]:\n",
    "                si = next(subject,None)\n",
    "            elif qi[2]<si[2]:\n",
    "                smry.append([False]*5+[qi[0],-1])\n",
    "                qi = next(query,None)\n",
    "            else:\n",
    "                #same pos has multiple snps\n",
    "                #query compare with each of them in subject\n",
    "                multi_snps.append(si)\n",
    "                smry.append(allele_match(qi[3],qi[4],si[3],si[4])+[qi[0],si[0]])\n",
    "                pre_s = si\n",
    "                si = next(subject,None)\n",
    "                if si is None or si[2] != pre_s[2]:\n",
    "                    qi = next(query,None)\n",
    "                    while(qi is not None and qi[2]==pre_s[2]):\n",
    "                        for s in multi_snps:\n",
    "                            smry.append(allele_match(qi[3],qi[4],s[3],s[4])+[qi[0],s[0]])\n",
    "                        qi = next(query,None)\n",
    "                    multi_snps = []\n",
    "                    \n",
    "    while(qi):\n",
    "        smry.append([False]*5+[qi[0],-1])\n",
    "        qi = next(query,None)\n",
    "    smry = pd.DataFrame(smry,columns=['keep','exact','flip','reverse','ambiguous','qidx','sidx'])\n",
    "    #print(smry.iloc[:,:5].groupby(['keep','exact','flip','reverse','ambiguous']))\n",
    "    if only_match:\n",
    "        smry = smry[smry.keep==True]\n",
    "    return smry\n",
    "def allele_match(a0,a1,ref0,ref1):\n",
    "    '''\n",
    "    input: a0 and a1 are the first snp, ref0 and ref1 are the second snp.\n",
    "    output: keep,exact,flip,reverse,both, complement. boolean values.\n",
    "    '''\n",
    "    keep,exact,flip,reverse=False,False,False,False\n",
    "     # reverse chain for multi base?\n",
    "    ca0,ca1 = Seq(a0).reverse_complement(),Seq(a1).reverse_complement()\n",
    "    if a0==ref0 and a1==ref1:\n",
    "        exact=True\n",
    "        keep=True\n",
    "    elif a0==ref1 and a1==ref0:\n",
    "        flip=True\n",
    "        keep=True\n",
    "        \n",
    "    if ca0==ref0 and ca1==ref1:\n",
    "        reverse = True\n",
    "        keep=True\n",
    "    elif ca0==ref1 and ca1==ref0:\n",
    "        flip=True\n",
    "        reverse = True\n",
    "        keep=True\n",
    "    return [keep,exact,flip,reverse,a0==ca1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "e2f081e8-30d0-48b5-9200-1b64c90356f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def check_ss1(ss,bim,keep_ambiguous=True):\n",
    "    '''index by chr+por+ordered ref and alt'''\n",
    "    ss.index = namebyordA0_A1(ss[['CHR','POS','A0','A1']])\n",
    "    bim.index = namebyordA0_A1(bim[['chrom','pos', 'a0', 'a1']])\n",
    "    # SNPs in bim should include all SNPs in ss\n",
    "    if(not ss.index.isin(bim.index).all()):\n",
    "        raise Exception(\"some SNPs in sumstat file are not in the bim file\")\n",
    "    #check duplicated index and remove them.\n",
    "    if(bim.index.duplicated().any() or ss.index.duplicated().any()):\n",
    "        warnings.warn(\"There are SNPs: REF:ALT = ALT:REF. They will be removed\")\n",
    "        bim = bim[~bim.index.duplicated(keep=False)]\n",
    "    bim = bim[bim.index.isin(ss.index)]\n",
    "    ss = ss.loc[bim.index]\n",
    "    if(ss.shape[0]!=bim.shape[0]):\n",
    "        raise Exception(\"Sumstas and genotype are not match with each other.\")\n",
    "    print(\"Paired SNPs\",ss.shape[0])\n",
    "    #input paired ss and bim\n",
    "    pm = pair_match(ss.A0,ss.A1,bim.a0,bim.a1)\n",
    "    if keep_ambiguous:\n",
    "        print('Warning: there are',sum(pm.ambiguous),'ambiguous SNPs')\n",
    "        pm = pm.iloc[:,1:]\n",
    "    else:\n",
    "        pm = pm[~pm.ambiguous].iloc[:,1:]\n",
    "    keep_idx = pm.any(axis=1)\n",
    "    keep_idx = keep_idx.index[keep_idx==True]\n",
    "    print(\"Overlap SNPs\",len(keep_idx))\n",
    "    #overlap snps by chr+pos+alleles.\n",
    "    new_subject = bim.loc[keep_idx][['chrom','pos', 'a0', 'a1','snp']]\n",
    "    new_subject.columns = ['CHR','POS','A0','A1','SNP']\n",
    "    #update beta and snp info\n",
    "    new_query = pd.concat([new_subject.iloc[:,:5],ss.loc[keep_idx].iloc[:,5:]],axis=1)\n",
    "    new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n",
    "    return new_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "3711f57b-e78c-4d0c-8001-dfb9b8e508e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def pair_match(a1,a2,ref1,ref2):\n",
    "    # a1 and a2 are the first data-set\n",
    "    # ref1 and ref2 are the 2nd data-set\n",
    "    # Make all the alleles into upper-case, as A,T,C,G:\n",
    "    a1 = a1.str.upper()\n",
    "    a2 = a2.str.upper()\n",
    "    ref1 = ref1.str.upper()\n",
    "    ref2 = ref2.str.upper()\n",
    "    # Strand flip, to change the allele representation in the 2nd data-set\n",
    "    flip1 = ref1.apply(strand_flip)\n",
    "    flip2 = ref2.apply(strand_flip)\n",
    "    result = {}\n",
    "    result[\"ambiguous\"] = ((a1==\"A\") & (a2==\"T\")) | ((a1==\"T\") & (a2==\"A\")) | ((a1==\"C\") & (a2==\"G\")) | ((a1==\"G\") & (a2==\"C\"))\n",
    "    # as long as scenario 1 is involved, sign_flip will return TRUE\n",
    "    result[\"sign_flip\"] = ((a1==ref2) & (a2==ref1)) | ((a1==flip2) & (a2==flip1))\n",
    "    # as long as scenario 2 is involved, strand_flip will return TRUE\n",
    "    result[\"strand_flip\"] = ((a1==flip1) & (a2==flip2)) | ((a1==flip2) & (a2==flip1))\n",
    "    # remove other cases, eg, tri-allelic, one dataset is A C, the other is A G, for example.\n",
    "    result[\"exact_match\"] = ((a1 == ref1) & (a2 == ref2))\n",
    "    return pd.DataFrame(result)\n",
    "def strand_flip(s):\n",
    "    return ''.join(Seq(s).reverse_complement())\n",
    "def namebyordA0_A1(df,cols=['CHR','POS','A0','A1']):\n",
    "    df.columns = cols\n",
    "    prefix = df[[x for x in cols if x not in ['CHR','POS','A0','A1']]+['CHR','POS']].astype(str).agg(':'.join, axis=1)\n",
    "    names = []\n",
    "    for p,A0,A1 in zip(prefix,df.A0,df.A1):\n",
    "        tmp = A0+':'+A1 if A0 > A1 else A1 +':'+ A0\n",
    "        names.append('_'.join([p,tmp]))\n",
    "    return names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "9170a078-5cee-40e3-a8c4-05cbdf48edfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def snps_match(query,subject,keep_ambiguous=True):\n",
    "    print(\"Total rows of query: \",query.shape[0],\"Total rows of subject: \",subject.shape[0])\n",
    "    if len(query.index[0].split('_')[0].split(':'))>2:\n",
    "        #gene:chr:pos case\n",
    "        genes_query = query.index.to_series().apply(lambda x: x.split('_')[0])\n",
    "        genes_subject = subject.index.to_series().apply(lambda x: x.split('_')[0])\n",
    "        query = dict(tuple(query.groupby(genes_query)))\n",
    "        subject = dict(tuple(subject.groupby(genes_subject)))\n",
    "        new_query, new_subject = [],[]\n",
    "        for g in genes_query.unique():\n",
    "            q = query.get(g)\n",
    "            s = subject.get(g)\n",
    "            if q is not None and s is not None:\n",
    "                new_q, new_s=snps_match_dup(q,s,keep_ambiguous)\n",
    "                new_query.append(new_q)\n",
    "                new_subject.append(new_s)\n",
    "        new_query, new_subject=pd.concat(new_query),pd.concat(new_query)\n",
    "    else:\n",
    "        #chr:pos case\n",
    "        new_query, new_subject=snps_match_dup(query,subject,keep_ambiguous)\n",
    "    return new_query, new_subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "c68dfe58-dda2-4246-a0f1-0caa03def07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "def snps_match_dup(query,subject,keep_ambiguous=True):\n",
    "    pm = compare_snps(query,subject)\n",
    "    if not keep_ambiguous:\n",
    "        pm = pm[~pm.ambiguous]\n",
    "    new_subject = subject.loc[pm.sidx]\n",
    "    #update beta and snp info\n",
    "    new_query = pd.concat([new_subject.iloc[:,:5],query.loc[pm.qidx].iloc[:,5:]],axis=1)\n",
    "    new_query.STAT[list(pm.flip)] = -new_query.STAT[list(pm.flip)]\n",
    "    return new_query, new_subject\n",
    "    \n",
    "def snps_match_nodup(query,subject,keep_ambiguous=True):\n",
    "    #overlap snps by chr+pos\n",
    "    print(\"Total rows of query: \",query.shape[0],\"Total rows of subject: \",subject.shape[0])\n",
    "    subject = subject[subject.index.isin(query.index)]\n",
    "    query = query.loc[subject.index]\n",
    "    print(\"Overlap chr:pos\",query.shape[0])\n",
    "    if query.index.duplicated().any():\n",
    "        raise Exception(\"There are duplicated chr:pos\")\n",
    "    pm = pair_match(query.A1,query.A0,subject.A1,subject.A0)\n",
    "    if keep_ambiguous:\n",
    "        print('Warning: there are',sum(pm.ambiguous),'ambiguous SNPs')\n",
    "        pm = pm.iloc[:,1:]\n",
    "    else:\n",
    "        pm = pm[~pm.ambiguous].iloc[:,1:]\n",
    "    keep_idx = pm.any(axis=1)\n",
    "    keep_idx = keep_idx.index[keep_idx==True]\n",
    "    print(\"Overlap SNPs\",len(keep_idx))\n",
    "    #overlap snps by chr+pos+alleles.\n",
    "    new_subject = subject.loc[keep_idx]\n",
    "    #update beta and snp info\n",
    "    new_query = pd.concat([new_subject.iloc[:,:5],query.loc[keep_idx].iloc[:,5:]],axis=1)\n",
    "    new_query.STAT[pm.sign_flip] = -new_query.STAT[pm.sign_flip]\n",
    "    return new_query,new_subject"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62570715-773b-4734-a4a1-2a47b4c4d464",
   "metadata": {},
   "source": [
    "## 2. Testing snps_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c188c15d-2893-404b-8d8c-f3198705fd37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fbb6784-1c4c-4c9a-9a9b-547b0f02c2a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6168835a-2a61-43df-a153-b7f2a2362cc1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a6515e10-4393-43c0-ace4-fd631b76f9a5",
   "metadata": {},
   "source": [
    "## 1.Testing check_ss1 function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b164a2df-01b2-4628-9f91-5ada5cf63e82",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "region = [5,272741,1213528]\n",
    "geno_path = '../MWE_region_extraction/ukb23156_c5.merged.filtered.5_272741_1213528.bed'\n",
    "sumstats_path = '../MWE_region_extraction/090321_UKBB_Hearing_aid_f3393_expandedwhite_6436cases_96601ctrl_PC1_2_f3393.regenie.snp_stats'\n",
    "pheno_path = None\n",
    "unr_path = 'MWE_region_extraction/UKB_genotypedatadownloaded083019.090221_sample_variant_qc_final_callrate90.filtered.extracted.white_europeans.filtered.092821_ldprun_unrelated.filtered.prune.txt'\n",
    "imp_geno_path = '../MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.bgen'\n",
    "imp_sumstats_path = '../MWE_region_extraction/100521_UKBB_Hearing_aid_f3393_expandedwhite_15601cases_237318ctrl_500k_PC1_PC2_f3393.regenie.snp_stats.gz'\n",
    "imp_ref = 'hg19'\n",
    "bgen_sample_path = '../MWE_region_extraction/ukb_imp_chr5_v3_05_272856_1213643.sample'\n",
    "output_sumstats = 'test.snp_stats.gz'\n",
    "output_LD = 'test_corr.csv.gz'\n",
    "\n",
    "#main(region,geno_path,sumstats_path,pheno_path,unr_path,imp_geno_path,imp_sumstats_path,imp_ref,output_sumstats,output_LD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dc0d8c05-1aef-43e8-b3eb-00c60af5f880",
   "metadata": {},
   "outputs": [],
   "source": [
    "imp_geno_path = '/mnt/mfs/statgen/archive/UKBiobank_Yale_transfer/ukb39554_imputeddataset/ukb_imp_chr5_v3.bgen'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "515fe087-9162-4ff9-bd10-a1dd5a6351fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this region [5, 272741, 1213528] has 3884 SNPs in Sumstat\n"
     ]
    }
   ],
   "source": [
    "exome_sumstats = Sumstat(sumstats_path)\n",
    "exome_sumstats.extractbyregion(region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "452dcb87-2084-43f7-9be9-fa14a775b5ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss.index = range(len(ss))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b2c4a3ef-1ca4-449f-b6b6-858fb46f2a6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5:272741_G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272741</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272741:A:G</td>\n",
       "      <td>1.789340</td>\n",
       "      <td>0.250526</td>\n",
       "      <td>7.055048e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272748_G:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272748</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272748:G:C</td>\n",
       "      <td>1.698750</td>\n",
       "      <td>0.279645</td>\n",
       "      <td>2.479648e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272755_G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272755</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272755:A:G</td>\n",
       "      <td>1.708990</td>\n",
       "      <td>0.242533</td>\n",
       "      <td>1.335826e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272758_G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272758</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272758:A:G</td>\n",
       "      <td>-1.285990</td>\n",
       "      <td>1.782120</td>\n",
       "      <td>4.705362e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272771_G:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272771</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272771:C:G</td>\n",
       "      <td>1.620460</td>\n",
       "      <td>0.466804</td>\n",
       "      <td>1.956681e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272791_C:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272791</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272791:A:C</td>\n",
       "      <td>0.014049</td>\n",
       "      <td>0.071979</td>\n",
       "      <td>8.452456e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272816_T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272816</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:272816:C:T</td>\n",
       "      <td>-0.259206</td>\n",
       "      <td>0.276487</td>\n",
       "      <td>3.485018e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272822_T:G</th>\n",
       "      <td>5</td>\n",
       "      <td>272822</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272822:T:G</td>\n",
       "      <td>-1.267800</td>\n",
       "      <td>1.258490</td>\n",
       "      <td>3.137451e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272829_T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272829</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272829:T:C</td>\n",
       "      <td>0.301320</td>\n",
       "      <td>0.878689</td>\n",
       "      <td>7.316595e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:275930_G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>275930</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:275930:G:A</td>\n",
       "      <td>-0.379434</td>\n",
       "      <td>0.533934</td>\n",
       "      <td>4.773083e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              CHR     POS REF ALT              SNP      BETA        SE  \\\n",
       "5:272741_G:A    5  272741   A   G  chr5:272741:A:G  1.789340  0.250526   \n",
       "5:272748_G:C    5  272748   G   C  chr5:272748:G:C  1.698750  0.279645   \n",
       "5:272755_G:A    5  272755   A   G  chr5:272755:A:G  1.708990  0.242533   \n",
       "5:272758_G:A    5  272758   A   G  chr5:272758:A:G -1.285990  1.782120   \n",
       "5:272771_G:C    5  272771   C   G  chr5:272771:C:G  1.620460  0.466804   \n",
       "5:272791_C:A    5  272791   A   C  chr5:272791:A:C  0.014049  0.071979   \n",
       "5:272816_T:C    5  272816   C   T  chr5:272816:C:T -0.259206  0.276487   \n",
       "5:272822_T:G    5  272822   T   G  chr5:272822:T:G -1.267800  1.258490   \n",
       "5:272829_T:C    5  272829   T   C  chr5:272829:T:C  0.301320  0.878689   \n",
       "5:275930_G:A    5  275930   G   A  chr5:275930:G:A -0.379434  0.533934   \n",
       "\n",
       "                         P  \n",
       "5:272741_G:A  7.055048e-11  \n",
       "5:272748_G:C  2.479648e-08  \n",
       "5:272755_G:A  1.335826e-10  \n",
       "5:272758_G:A  4.705362e-01  \n",
       "5:272771_G:C  1.956681e-03  \n",
       "5:272791_C:A  8.452456e-01  \n",
       "5:272816_T:C  3.485018e-01  \n",
       "5:272822_T:G  3.137451e-01  \n",
       "5:272829_T:C  7.316595e-01  \n",
       "5:275930_G:A  4.773083e-01  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss.loc[ss.index[:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "86c7af0f-298b-4722-8fa2-9f01b31fe4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = namebyordA0_A1(exome_sumstats.ss[['CHR','POS','REF','ALT']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "87361364-1f70-4566-9394-8ed285de080d",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = pd.Series(tmp).str.split('_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "5d55b7d0-7625-4f8e-80f6-1e50ad501200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        [5:272741, G:A]\n",
       "1        [5:272748, G:C]\n",
       "2        [5:272755, G:A]\n",
       "3        [5:272758, G:A]\n",
       "4        [5:272771, G:C]\n",
       "              ...       \n",
       "3879    [5:1213517, T:C]\n",
       "3880    [5:1213518, G:A]\n",
       "3881    [5:1213524, T:C]\n",
       "3882    [5:1213527, T:C]\n",
       "3883    [5:1213528, G:A]\n",
       "Length: 3884, dtype: object"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "75daae8b-5cd5-444f-accb-6082e8dc0fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss = exome_sumstats.ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f46564b3-e94a-49f2-877c-b37be184b9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss.index = pd.Series(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0f309ae7-45e6-47e4-aa43-b89072f92fb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['5:272741_G:A', '5:272748_G:C', '5:272755_G:A', '5:272758_G:A',\n",
       "       '5:272771_G:C', '5:272791_C:A', '5:272816_T:C', '5:272822_T:G',\n",
       "       '5:272829_T:C', '5:275930_G:A',\n",
       "       ...\n",
       "       '5:1213465_T:C', '5:1213471_T:C', '5:1213482_T:C', '5:1213483_G:A',\n",
       "       '5:1213495_T:C', '5:1213517_T:C', '5:1213518_G:A', '5:1213524_T:C',\n",
       "       '5:1213527_T:C', '5:1213528_G:A'],\n",
       "      dtype='object', length=3884)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "866ce2d6-062d-4285-882f-4cd96805e73d",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = ss.index.to_series().apply(lambda x: x.split(':')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "335f44d8-7a24-4ea1-b195-9d205624561d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['5'], dtype=object)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "dddbdec4-cff0-49b0-b65f-6f6f4c2f16e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ss.index[0].split('_')[0].split(':'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "fe7d812e-9f89-4a16-b7ee-df0b9e8276cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ab'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'ab_c'.split('_')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "cac08995-66cd-45af-81c3-1ff210d1a604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, ...,  True,  True,  True])"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss.index.isin(ss.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "9a5a4d99-5770-47e3-90de-37d3b2484c23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['5:432782_TC:T', '5:432782_TC:T', '5:473309_AGCG:A', '5:473309_AGCG:A',\n",
       "       '5:612613_AC:A', '5:612613_AC:A', '5:665193_GCCTTGC:G',\n",
       "       '5:665193_GCCTTGC:G', '5:692497_GC:G', '5:692497_GC:G', '5:889546_AC:A',\n",
       "       '5:889546_AC:A', '5:891687_ACTT:A', '5:891687_ACTT:A', '5:912005_AT:A',\n",
       "       '5:912005_AT:A', '5:1035376_AC:A', '5:1035376_AC:A',\n",
       "       '5:1038307_GCACGAGCACCACCACCAC:G', '5:1038307_GCACGAGCACCACCACCAC:G',\n",
       "       '5:1038331_GCAC:G', '5:1038331_GCACCAC:G', '5:1038331_GCAC:G',\n",
       "       '5:1038331_GCACCACCAC:G', '5:1038331_GCACCAC:G',\n",
       "       '5:1038331_GCACCACCAC:G', '5:1056514_CCA:C', '5:1056514_CCA:C',\n",
       "       '5:1112017_TCCCGG:T', '5:1112017_TCCCGG:T'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss.index[ss.index.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e139d435-53d2-44e8-8d00-cbe9664c0c17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['5:272741', 'G:A'],\n",
       " ['5:272748', 'G:C'],\n",
       " ['5:272755', 'G:A'],\n",
       " ['5:272758', 'G:A'],\n",
       " ['5:272771', 'G:C'],\n",
       " ['5:272791', 'C:A'],\n",
       " ['5:272816', 'T:C'],\n",
       " ['5:272822', 'T:G'],\n",
       " ['5:272829', 'T:C'],\n",
       " ['5:275930', 'G:A']]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "45ae5946-b2ac-4c4a-be95-f9bb78b5b640",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    5:272741\n",
       "1    5:272748\n",
       "2    5:272755\n",
       "3    5:272758\n",
       "4    5:272771\n",
       "5    5:272791\n",
       "6    5:272816\n",
       "7    5:272822\n",
       "8    5:272829\n",
       "9    5:275930\n",
       "dtype: object"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.apply(lambda x: x[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b467afed-4cea-4fa5-8dfc-3da8cc695c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "exome_geno = Genodata(geno_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ff07d687-5b15-4b4f-b817-078b8e1e8456",
   "metadata": {},
   "outputs": [],
   "source": [
    "bim = exome_geno.bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "801e42cb-edbf-4826-9b4b-a9efe10c0d69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paired SNPs 3847\n",
      "Warning: there are 402 ambiguous SNPs\n",
      "Overlap SNPs 3847\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/2206559.1.plot.q/ipykernel_14021/227429081.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5:272741:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272741</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:272741:G:A</td>\n",
       "      <td>-1.789340</td>\n",
       "      <td>0.250526</td>\n",
       "      <td>7.055048e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272748:G:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272748</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272748:C:G</td>\n",
       "      <td>-1.698750</td>\n",
       "      <td>0.279645</td>\n",
       "      <td>2.479648e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272755:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272755</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:272755:G:A</td>\n",
       "      <td>-1.708990</td>\n",
       "      <td>0.242533</td>\n",
       "      <td>1.335826e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272758:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>272758</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:272758:G:A</td>\n",
       "      <td>1.285990</td>\n",
       "      <td>1.782120</td>\n",
       "      <td>4.705362e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:272771:G:C</th>\n",
       "      <td>5</td>\n",
       "      <td>272771</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272771:G:C</td>\n",
       "      <td>-1.620460</td>\n",
       "      <td>0.466804</td>\n",
       "      <td>1.956681e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:1213517:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>1213517</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:1213517:T:C</td>\n",
       "      <td>1.093870</td>\n",
       "      <td>1.988000</td>\n",
       "      <td>5.821595e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:1213518:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>1213518</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:1213518:A:G</td>\n",
       "      <td>1.221880</td>\n",
       "      <td>1.772910</td>\n",
       "      <td>4.906999e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:1213524:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>1213524</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213524:C:T</td>\n",
       "      <td>-0.151751</td>\n",
       "      <td>0.664055</td>\n",
       "      <td>8.192405e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:1213527:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>1213527</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:1213527:T:C</td>\n",
       "      <td>1.080770</td>\n",
       "      <td>1.865680</td>\n",
       "      <td>5.623918e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:1213528:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>1213528</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:1213528:A:G</td>\n",
       "      <td>1.057160</td>\n",
       "      <td>2.654680</td>\n",
       "      <td>6.904639e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3847 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CHR      POS REF ALT               SNP      BETA        SE  \\\n",
       "5:272741:G:A     5   272741   G   A   chr5:272741:G:A -1.789340  0.250526   \n",
       "5:272748:G:C     5   272748   C   G   chr5:272748:C:G -1.698750  0.279645   \n",
       "5:272755:G:A     5   272755   G   A   chr5:272755:G:A -1.708990  0.242533   \n",
       "5:272758:G:A     5   272758   G   A   chr5:272758:G:A  1.285990  1.782120   \n",
       "5:272771:G:C     5   272771   G   C   chr5:272771:G:C -1.620460  0.466804   \n",
       "...            ...      ...  ..  ..               ...       ...       ...   \n",
       "5:1213517:T:C    5  1213517   T   C  chr5:1213517:T:C  1.093870  1.988000   \n",
       "5:1213518:G:A    5  1213518   A   G  chr5:1213518:A:G  1.221880  1.772910   \n",
       "5:1213524:T:C    5  1213524   C   T  chr5:1213524:C:T -0.151751  0.664055   \n",
       "5:1213527:T:C    5  1213527   T   C  chr5:1213527:T:C  1.080770  1.865680   \n",
       "5:1213528:G:A    5  1213528   A   G  chr5:1213528:A:G  1.057160  2.654680   \n",
       "\n",
       "                          P  \n",
       "5:272741:G:A   7.055048e-11  \n",
       "5:272748:G:C   2.479648e-08  \n",
       "5:272755:G:A   1.335826e-10  \n",
       "5:272758:G:A   4.705362e-01  \n",
       "5:272771:G:C   1.956681e-03  \n",
       "...                     ...  \n",
       "5:1213517:T:C  5.821595e-01  \n",
       "5:1213518:G:A  4.906999e-01  \n",
       "5:1213524:T:C  8.192405e-01  \n",
       "5:1213527:T:C  5.623918e-01  \n",
       "5:1213528:G:A  6.904639e-01  \n",
       "\n",
       "[3847 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_ss1(exome_sumstats.ss,bim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c382fdd8-d56b-49a6-8fd6-7e0db5667f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "imput_sumstats = Sumstat(imp_sumstats_path)\n",
    "imput_geno = Genodata(imp_geno_path,bgen_sample_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "83cca0ec-4c25-4bd3-bec3-54bbe92efbf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Paired SNPs 632\n",
      "Warning: there are 85 ambiguous SNPs\n",
      "Overlap SNPs 632\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/2206559.1.plot.q/ipykernel_3137/3698415862.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5:73072354:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>73072354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73072354:T:C</td>\n",
       "      <td>0.101991</td>\n",
       "      <td>0.012277</td>\n",
       "      <td>1.073001e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072504:C:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072504</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73072504:A:C</td>\n",
       "      <td>0.015024</td>\n",
       "      <td>0.036310</td>\n",
       "      <td>6.790301e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072530:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072530</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73072530:A:G</td>\n",
       "      <td>0.095930</td>\n",
       "      <td>0.050772</td>\n",
       "      <td>5.883557e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072863:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072863</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:73072863:G:A</td>\n",
       "      <td>-0.072351</td>\n",
       "      <td>0.667580</td>\n",
       "      <td>9.136960e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072873:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072873</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73072873:A:G</td>\n",
       "      <td>0.042773</td>\n",
       "      <td>0.543868</td>\n",
       "      <td>9.373146e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144301:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>73144301</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73144301:T:C</td>\n",
       "      <td>0.079786</td>\n",
       "      <td>0.012277</td>\n",
       "      <td>8.143292e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144431:TAG:T</th>\n",
       "      <td>5</td>\n",
       "      <td>73144431</td>\n",
       "      <td>TAG</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:73144431:TAG:T</td>\n",
       "      <td>0.029964</td>\n",
       "      <td>0.023528</td>\n",
       "      <td>2.028154e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144432:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144432</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73144432:A:G</td>\n",
       "      <td>0.088728</td>\n",
       "      <td>0.012569</td>\n",
       "      <td>1.627796e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144716:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144716</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:73144716:G:A</td>\n",
       "      <td>0.300375</td>\n",
       "      <td>0.575205</td>\n",
       "      <td>6.015282e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144845:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144845</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73144845:A:G</td>\n",
       "      <td>0.090138</td>\n",
       "      <td>0.012494</td>\n",
       "      <td>5.022269e-13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CHR       POS  REF ALT                  SNP      BETA  \\\n",
       "5:73072354:T:C      5  73072354    T   C    chr5:73072354:T:C  0.101991   \n",
       "5:73072504:C:A      5  73072504    A   C    chr5:73072504:A:C  0.015024   \n",
       "5:73072530:G:A      5  73072530    A   G    chr5:73072530:A:G  0.095930   \n",
       "5:73072863:G:A      5  73072863    G   A    chr5:73072863:G:A -0.072351   \n",
       "5:73072873:G:A      5  73072873    A   G    chr5:73072873:A:G  0.042773   \n",
       "...               ...       ...  ...  ..                  ...       ...   \n",
       "5:73144301:T:C      5  73144301    T   C    chr5:73144301:T:C  0.079786   \n",
       "5:73144431:TAG:T    5  73144431  TAG   T  chr5:73144431:TAG:T  0.029964   \n",
       "5:73144432:G:A      5  73144432    A   G    chr5:73144432:A:G  0.088728   \n",
       "5:73144716:G:A      5  73144716    G   A    chr5:73144716:G:A  0.300375   \n",
       "5:73144845:G:A      5  73144845    A   G    chr5:73144845:A:G  0.090138   \n",
       "\n",
       "                        SE             P  \n",
       "5:73072354:T:C    0.012277  1.073001e-16  \n",
       "5:73072504:C:A    0.036310  6.790301e-01  \n",
       "5:73072530:G:A    0.050772  5.883557e-02  \n",
       "5:73072863:G:A    0.667580  9.136960e-01  \n",
       "5:73072873:G:A    0.543868  9.373146e-01  \n",
       "...                    ...           ...  \n",
       "5:73144301:T:C    0.012277  8.143292e-11  \n",
       "5:73144431:TAG:T  0.023528  2.028154e-01  \n",
       "5:73144432:G:A    0.012569  1.627796e-12  \n",
       "5:73144716:G:A    0.575205  6.015282e-01  \n",
       "5:73144845:G:A    0.012494  5.022269e-13  \n",
       "\n",
       "[632 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_ss1(imput_sumstats.ss,imput_geno.bim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "aab48b5c-24f8-4a9d-8b7b-354ae315e10f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5:73072354:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>73072354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73072354:T:C</td>\n",
       "      <td>0.101991</td>\n",
       "      <td>0.012277</td>\n",
       "      <td>1.073001e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072504:C:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072504</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73072504:A:C</td>\n",
       "      <td>0.015024</td>\n",
       "      <td>0.036310</td>\n",
       "      <td>6.790301e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072530:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072530</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73072530:A:G</td>\n",
       "      <td>0.095930</td>\n",
       "      <td>0.050772</td>\n",
       "      <td>5.883557e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072863:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072863</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:73072863:G:A</td>\n",
       "      <td>-0.072351</td>\n",
       "      <td>0.667580</td>\n",
       "      <td>9.136960e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072873:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73072873</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73072873:A:G</td>\n",
       "      <td>0.042773</td>\n",
       "      <td>0.543868</td>\n",
       "      <td>9.373146e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144301:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>73144301</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:73144301:T:C</td>\n",
       "      <td>0.079786</td>\n",
       "      <td>0.012277</td>\n",
       "      <td>8.143292e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144431:TAG:T</th>\n",
       "      <td>5</td>\n",
       "      <td>73144431</td>\n",
       "      <td>TAG</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:73144431:TAG:T</td>\n",
       "      <td>0.029964</td>\n",
       "      <td>0.023528</td>\n",
       "      <td>2.028154e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144432:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144432</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73144432:A:G</td>\n",
       "      <td>0.088728</td>\n",
       "      <td>0.012569</td>\n",
       "      <td>1.627796e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144716:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144716</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:73144716:G:A</td>\n",
       "      <td>0.300375</td>\n",
       "      <td>0.575205</td>\n",
       "      <td>6.015282e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144845:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>73144845</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:73144845:A:G</td>\n",
       "      <td>0.090138</td>\n",
       "      <td>0.012494</td>\n",
       "      <td>5.022269e-13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>632 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CHR       POS  REF ALT                  SNP      BETA  \\\n",
       "5:73072354:T:C      5  73072354    T   C    chr5:73072354:T:C  0.101991   \n",
       "5:73072504:C:A      5  73072504    A   C    chr5:73072504:A:C  0.015024   \n",
       "5:73072530:G:A      5  73072530    A   G    chr5:73072530:A:G  0.095930   \n",
       "5:73072863:G:A      5  73072863    G   A    chr5:73072863:G:A -0.072351   \n",
       "5:73072873:G:A      5  73072873    A   G    chr5:73072873:A:G  0.042773   \n",
       "...               ...       ...  ...  ..                  ...       ...   \n",
       "5:73144301:T:C      5  73144301    T   C    chr5:73144301:T:C  0.079786   \n",
       "5:73144431:TAG:T    5  73144431  TAG   T  chr5:73144431:TAG:T  0.029964   \n",
       "5:73144432:G:A      5  73144432    A   G    chr5:73144432:A:G  0.088728   \n",
       "5:73144716:G:A      5  73144716    G   A    chr5:73144716:G:A  0.300375   \n",
       "5:73144845:G:A      5  73144845    A   G    chr5:73144845:A:G  0.090138   \n",
       "\n",
       "                        SE             P  \n",
       "5:73072354:T:C    0.012277  1.073001e-16  \n",
       "5:73072504:C:A    0.036310  6.790301e-01  \n",
       "5:73072530:G:A    0.050772  5.883557e-02  \n",
       "5:73072863:G:A    0.667580  9.136960e-01  \n",
       "5:73072873:G:A    0.543868  9.373146e-01  \n",
       "...                    ...           ...  \n",
       "5:73144301:T:C    0.012277  8.143292e-11  \n",
       "5:73144431:TAG:T  0.023528  2.028154e-01  \n",
       "5:73144432:G:A    0.012569  1.627796e-12  \n",
       "5:73144716:G:A    0.575205  6.015282e-01  \n",
       "5:73144845:G:A    0.012494  5.022269e-13  \n",
       "\n",
       "[632 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e7b3cc25-05a1-466b-9b0e-4c5ff6cee439",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>snp</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>a0</th>\n",
       "      <th>a1</th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5:73072354:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73072354:T:C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73072354</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2401738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072371:T:C</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73072371:C:T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73072371</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2401739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072425:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73072425:G:A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73072425</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2401740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072501:T:G</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73072501:G:T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73072501</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>2401741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73072504:C:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73072504:A:C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73072504</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2401742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144742:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73144742:G:A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73144742</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2404138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144750:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73144750:A:G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73144750</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2404139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144765:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73144765:A:G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73144765</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2404140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144823:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73144823:G:A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73144823</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2404141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5:73144845:G:A</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:73144845:A:G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>73144845</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2404142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2405 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                chrom                snp   cm       pos a0 a1        i\n",
       "5:73072354:T:C      5  chr5:73072354:T:C  0.0  73072354  T  C  2401738\n",
       "5:73072371:T:C      5  chr5:73072371:C:T  0.0  73072371  C  T  2401739\n",
       "5:73072425:G:A      5  chr5:73072425:G:A  0.0  73072425  G  A  2401740\n",
       "5:73072501:T:G      5  chr5:73072501:G:T  0.0  73072501  G  T  2401741\n",
       "5:73072504:C:A      5  chr5:73072504:A:C  0.0  73072504  A  C  2401742\n",
       "...               ...                ...  ...       ... .. ..      ...\n",
       "5:73144742:G:A      5  chr5:73144742:G:A  0.0  73144742  G  A  2404138\n",
       "5:73144750:G:A      5  chr5:73144750:A:G  0.0  73144750  A  G  2404139\n",
       "5:73144765:G:A      5  chr5:73144765:A:G  0.0  73144765  A  G  2404140\n",
       "5:73144823:G:A      5  chr5:73144823:G:A  0.0  73144823  G  A  2404141\n",
       "5:73144845:G:A      5  chr5:73144845:A:G  0.0  73144845  A  G  2404142\n",
       "\n",
       "[2405 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4d5bfd5c-69e2-41d4-897b-4f19df47f80f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sumstat:                  CHR       POS  REF ALT                  SNP      BETA  \\\n",
       "5:73072354:T:C      5  73072354    T   C    chr5:73072354:T:C  0.101991   \n",
       "5:73072504:C:A      5  73072504    A   C    chr5:73072504:A:C  0.015024   \n",
       "5:73072530:G:A      5  73072530    A   G    chr5:73072530:A:G  0.095930   \n",
       "5:73072863:G:A      5  73072863    G   A    chr5:73072863:G:A -0.072351   \n",
       "5:73072873:G:A      5  73072873    A   G    chr5:73072873:A:G  0.042773   \n",
       "...               ...       ...  ...  ..                  ...       ...   \n",
       "5:73144301:T:C      5  73144301    T   C    chr5:73144301:T:C  0.079786   \n",
       "5:73144431:TAG:T    5  73144431  TAG   T  chr5:73144431:TAG:T  0.029964   \n",
       "5:73144432:G:A      5  73144432    A   G    chr5:73144432:A:G  0.088728   \n",
       "5:73144716:G:A      5  73144716    G   A    chr5:73144716:G:A  0.300375   \n",
       "5:73144845:G:A      5  73144845    A   G    chr5:73144845:A:G  0.090138   \n",
       "\n",
       "                        SE             P  \n",
       "5:73072354:T:C    0.012277  1.073001e-16  \n",
       "5:73072504:C:A    0.036310  6.790301e-01  \n",
       "5:73072530:G:A    0.050772  5.883557e-02  \n",
       "5:73072863:G:A    0.667580  9.136960e-01  \n",
       "5:73072873:G:A    0.543868  9.373146e-01  \n",
       "...                    ...           ...  \n",
       "5:73144301:T:C    0.012277  8.143292e-11  \n",
       "5:73144431:TAG:T  0.023528  2.028154e-01  \n",
       "5:73144432:G:A    0.012569  1.627796e-12  \n",
       "5:73144716:G:A    0.575205  6.015282e-01  \n",
       "5:73144845:G:A    0.012494  5.022269e-13  \n",
       "\n",
       "[632 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imput_sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "256279b6-1af8-4c9f-bbc2-58143f9a89a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this region (5, 73072354, 73144845) has 632 SNPs in Sumstat\n",
      "this region (5, 73072354, 73144845) has 2405 SNPs in Genodata\n",
      "Paired SNPs 632\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Can only compare identically-labeled Series objects",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/2206559.1.plot.q/ipykernel_3137/2846184474.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mimput_sumstats\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextractbyregion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimp_region\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mimput_geno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextractbyregion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimp_region\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mimput_sumstats\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmatch_ss\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimput_geno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbim\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0mimput_geno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgeno_in_stat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimput_sumstats\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mss\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/mfs/statgen/yin/Github/LDtools/LDtools/sumstat.py\u001b[0m in \u001b[0;36mmatch_ss\u001b[0;34m(self, bim)\u001b[0m\n\u001b[1;32m     61\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmatch_ss\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbim\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mss\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_ss1\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mss\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbim\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/mfs/statgen/yin/Github/LDtools/LDtools/utils.py\u001b[0m in \u001b[0;36mcheck_ss1\u001b[0;34m(ss, bim, keep_ambiguous)\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Paired SNPs\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m     \u001b[0;31m#input paired ss and bim\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 155\u001b[0;31m     \u001b[0mpm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpair_match\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mREF\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mALT\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0ma0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    156\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mkeep_ambiguous\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    157\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Warning: there are'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mambiguous\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'ambiguous SNPs'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/mfs/statgen/yin/Github/LDtools/LDtools/utils.py\u001b[0m in \u001b[0;36mpair_match\u001b[0;34m(a1, a2, ref1, ref2)\u001b[0m\n\u001b[1;32m    183\u001b[0m     \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ambiguous\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"A\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"T\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"T\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"A\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"C\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"G\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"G\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"C\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    184\u001b[0m     \u001b[0;31m# as long as scenario 1 is involved, sign_flip will return TRUE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 185\u001b[0;31m     \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"sign_flip\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mref2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mref1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    186\u001b[0m     \u001b[0;31m# as long as scenario 2 is involved, strand_flip will return TRUE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    187\u001b[0m     \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"strand_flip\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip1\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma1\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma2\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mflip1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/ops/common.py\u001b[0m in \u001b[0;36mnew_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     67\u001b[0m         \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitem_from_zerodim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mnew_method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/arraylike.py\u001b[0m in \u001b[0;36m__eq__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     30\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__eq__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__eq__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cmp_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moperator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0munpack_zerodim_and_defer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"__ne__\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_cmp_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   5494\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5495\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_indexed_same\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5496\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Can only compare identically-labeled Series objects\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5497\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5498\u001b[0m         \u001b[0mlvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Can only compare identically-labeled Series objects"
     ]
    }
   ],
   "source": [
    "region = [5, 73776529, 73849020]\n",
    "hg38toimpref = Liftover('hg38','hg19')\n",
    "imp_region = hg38toimpref.region_liftover(region)\n",
    "imput_sumstats.extractbyregion(imp_region)\n",
    "imput_geno.extractbyregion(imp_region)\n",
    "imput_sumstats.match_ss(imput_geno.bim)\n",
    "imput_geno.geno_in_stat(imput_sumstats.ss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13f0e857-d4ff-45b8-b705-1a91ff855bc7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9807a33d-a013-42ac-b367-9d52eb262207",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "a=exome_sumstats.sample(n=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "a8e44ea7-02f4-4191-a007-404317a7fd5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = exome_sumstats.ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "720a6369-7bfa-4553-8ab2-a24bc06e435e",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = a.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "d361ddd3-7e30-416a-b264-5c6aec15c641",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "aa = a.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "5389b054-9039-49d1-8352-940b94efd9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa.REF = list(a.ALT)\n",
    "aa.ALT = list(a.REF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "8c0365d3-32a7-4b57-84ac-2644682bd6cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "keep   exact  flip   reverse  both   complement\n",
      "False  False  False  False    False  False         8439\n",
      "True   False  False  True     False  False           13\n",
      "       True   False  False    False  False            9\n",
      "       False  True   False    False  False            5\n",
      "False  False  False  False    False  True             4\n",
      "True   False  False  False    True   False            2\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "tmp =compare_snps(imput_sumstats.ss,exome_sumstats.ss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "b44e1ad8-36b5-4a26-98c7-68dfdf6a34fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keep</th>\n",
       "      <th>exact</th>\n",
       "      <th>flip</th>\n",
       "      <th>reverse</th>\n",
       "      <th>both</th>\n",
       "      <th>complement</th>\n",
       "      <th>qidx</th>\n",
       "      <th>sidx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6767726</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6767727</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6767728</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6767729</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6767730</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8467</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6776191</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8468</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6776192</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8469</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6776193</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8470</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6776194</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8471</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>6776195</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8472 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       keep  exact   flip  reverse   both  complement     qidx  sidx\n",
       "0     False  False  False    False  False       False  6767726    -1\n",
       "1     False  False  False    False  False       False  6767727    -1\n",
       "2     False  False  False    False  False       False  6767728    -1\n",
       "3     False  False  False    False  False       False  6767729    -1\n",
       "4     False  False  False    False  False       False  6767730    -1\n",
       "...     ...    ...    ...      ...    ...         ...      ...   ...\n",
       "8467  False  False  False    False  False       False  6776191    -1\n",
       "8468  False  False  False    False  False       False  6776192    -1\n",
       "8469  False  False  False    False  False       False  6776193    -1\n",
       "8470  False  False  False    False  False       False  6776194    -1\n",
       "8471  False  False  False    False  False       False  6776195    -1\n",
       "\n",
       "[8472 rows x 8 columns]"
      ]
     },
     "execution_count": 302,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "882bd97e-0d51-494e-be72-9efcb5900689",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6767726</th>\n",
       "      <td>5</td>\n",
       "      <td>272851</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272851:A:G</td>\n",
       "      <td>0.357496</td>\n",
       "      <td>0.888197</td>\n",
       "      <td>0.687318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767727</th>\n",
       "      <td>5</td>\n",
       "      <td>272906</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272906:A:C</td>\n",
       "      <td>-0.003007</td>\n",
       "      <td>0.019764</td>\n",
       "      <td>0.879070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767728</th>\n",
       "      <td>5</td>\n",
       "      <td>273143</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:273143:A:G</td>\n",
       "      <td>-0.013693</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0.412684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767729</th>\n",
       "      <td>5</td>\n",
       "      <td>273160</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:273160:G:C</td>\n",
       "      <td>0.235713</td>\n",
       "      <td>0.348772</td>\n",
       "      <td>0.499145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767730</th>\n",
       "      <td>5</td>\n",
       "      <td>273534</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:273534:C:T</td>\n",
       "      <td>0.050095</td>\n",
       "      <td>0.139496</td>\n",
       "      <td>0.719509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776191</th>\n",
       "      <td>5</td>\n",
       "      <td>1213094</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213094:C:T</td>\n",
       "      <td>-0.015881</td>\n",
       "      <td>0.023298</td>\n",
       "      <td>0.495462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776192</th>\n",
       "      <td>5</td>\n",
       "      <td>1213134</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:1213134:G:A</td>\n",
       "      <td>-1.142280</td>\n",
       "      <td>1.344380</td>\n",
       "      <td>0.395509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776193</th>\n",
       "      <td>5</td>\n",
       "      <td>1213223</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213223:C:T</td>\n",
       "      <td>-0.003009</td>\n",
       "      <td>0.013631</td>\n",
       "      <td>0.825270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776194</th>\n",
       "      <td>5</td>\n",
       "      <td>1213404</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>chr5:1213404:T:TC</td>\n",
       "      <td>-0.039146</td>\n",
       "      <td>0.117837</td>\n",
       "      <td>0.739735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776195</th>\n",
       "      <td>5</td>\n",
       "      <td>1213510</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213510:C:T</td>\n",
       "      <td>0.009318</td>\n",
       "      <td>0.012922</td>\n",
       "      <td>0.470845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8461 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR      POS REF ALT                SNP      BETA        SE         P\n",
       "6767726    5   272851   A   G    chr5:272851:A:G  0.357496  0.888197  0.687318\n",
       "6767727    5   272906   A   C    chr5:272906:A:C -0.003007  0.019764  0.879070\n",
       "6767728    5   273143   A   G    chr5:273143:A:G -0.013693  0.016716  0.412684\n",
       "6767729    5   273160   G   C    chr5:273160:G:C  0.235713  0.348772  0.499145\n",
       "6767730    5   273534   C   T    chr5:273534:C:T  0.050095  0.139496  0.719509\n",
       "...      ...      ...  ..  ..                ...       ...       ...       ...\n",
       "6776191    5  1213094   C   T   chr5:1213094:C:T -0.015881  0.023298  0.495462\n",
       "6776192    5  1213134   G   A   chr5:1213134:G:A -1.142280  1.344380  0.395509\n",
       "6776193    5  1213223   C   T   chr5:1213223:C:T -0.003009  0.013631  0.825270\n",
       "6776194    5  1213404   T  TC  chr5:1213404:T:TC -0.039146  0.117837  0.739735\n",
       "6776195    5  1213510   C   T   chr5:1213510:C:T  0.009318  0.012922  0.470845\n",
       "\n",
       "[8461 rows x 8 columns]"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imput_sumstats.ss.loc[tmp.qidx[tmp.exact==False].drop_duplicates()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "0ad56f0e-6b9e-46cb-ba99-5aa8c64d3dfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6767726</th>\n",
       "      <td>5</td>\n",
       "      <td>272851</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:272851:A:G</td>\n",
       "      <td>0.357496</td>\n",
       "      <td>0.888197</td>\n",
       "      <td>0.687318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767727</th>\n",
       "      <td>5</td>\n",
       "      <td>272906</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:272906:A:C</td>\n",
       "      <td>-0.003007</td>\n",
       "      <td>0.019764</td>\n",
       "      <td>0.879070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767728</th>\n",
       "      <td>5</td>\n",
       "      <td>273143</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:273143:A:G</td>\n",
       "      <td>-0.013693</td>\n",
       "      <td>0.016716</td>\n",
       "      <td>0.412684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767729</th>\n",
       "      <td>5</td>\n",
       "      <td>273160</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5:273160:G:C</td>\n",
       "      <td>0.235713</td>\n",
       "      <td>0.348772</td>\n",
       "      <td>0.499145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6767730</th>\n",
       "      <td>5</td>\n",
       "      <td>273534</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:273534:C:T</td>\n",
       "      <td>0.050095</td>\n",
       "      <td>0.139496</td>\n",
       "      <td>0.719509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776191</th>\n",
       "      <td>5</td>\n",
       "      <td>1213094</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213094:C:T</td>\n",
       "      <td>-0.015881</td>\n",
       "      <td>0.023298</td>\n",
       "      <td>0.495462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776192</th>\n",
       "      <td>5</td>\n",
       "      <td>1213134</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:1213134:G:A</td>\n",
       "      <td>-1.142280</td>\n",
       "      <td>1.344380</td>\n",
       "      <td>0.395509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776193</th>\n",
       "      <td>5</td>\n",
       "      <td>1213223</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213223:C:T</td>\n",
       "      <td>-0.003009</td>\n",
       "      <td>0.013631</td>\n",
       "      <td>0.825270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776194</th>\n",
       "      <td>5</td>\n",
       "      <td>1213404</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>chr5:1213404:T:TC</td>\n",
       "      <td>-0.039146</td>\n",
       "      <td>0.117837</td>\n",
       "      <td>0.739735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776195</th>\n",
       "      <td>5</td>\n",
       "      <td>1213510</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:1213510:C:T</td>\n",
       "      <td>0.009318</td>\n",
       "      <td>0.012922</td>\n",
       "      <td>0.470845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8463 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR      POS REF ALT                SNP      BETA        SE         P\n",
       "6767726    5   272851   A   G    chr5:272851:A:G  0.357496  0.888197  0.687318\n",
       "6767727    5   272906   A   C    chr5:272906:A:C -0.003007  0.019764  0.879070\n",
       "6767728    5   273143   A   G    chr5:273143:A:G -0.013693  0.016716  0.412684\n",
       "6767729    5   273160   G   C    chr5:273160:G:C  0.235713  0.348772  0.499145\n",
       "6767730    5   273534   C   T    chr5:273534:C:T  0.050095  0.139496  0.719509\n",
       "...      ...      ...  ..  ..                ...       ...       ...       ...\n",
       "6776191    5  1213094   C   T   chr5:1213094:C:T -0.015881  0.023298  0.495462\n",
       "6776192    5  1213134   G   A   chr5:1213134:G:A -1.142280  1.344380  0.395509\n",
       "6776193    5  1213223   C   T   chr5:1213223:C:T -0.003009  0.013631  0.825270\n",
       "6776194    5  1213404   T  TC  chr5:1213404:T:TC -0.039146  0.117837  0.739735\n",
       "6776195    5  1213510   C   T   chr5:1213510:C:T  0.009318  0.012922  0.470845\n",
       "\n",
       "[8463 rows x 8 columns]"
      ]
     },
     "execution_count": 305,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imput_sumstats.ss.loc[tmp.qidx[tmp.exact==False]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "09885a8d-c22b-4080-b444-2658ac8993cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "keep   exact  flip   reverse  both   complement\n",
      "False  False  False  False    False  False         118\n",
      "True   False  True   False    False  False           1\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(tmp.iloc[:,:6].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "6f2fcade-bbbb-49a3-b73d-df5ced72b2c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 263,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(tmp['query']==-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "6f0b5125-4c16-4f9a-aeec-a78f93c9519c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keep</th>\n",
       "      <th>exact</th>\n",
       "      <th>flip</th>\n",
       "      <th>reverse</th>\n",
       "      <th>both</th>\n",
       "      <th>complement</th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811358</td>\n",
       "      <td>811358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811368</td>\n",
       "      <td>811368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811381</td>\n",
       "      <td>811381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811395</td>\n",
       "      <td>811395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811400</td>\n",
       "      <td>811400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811402</td>\n",
       "      <td>811402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811415</td>\n",
       "      <td>811415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811422</td>\n",
       "      <td>811422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811423</td>\n",
       "      <td>811423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811431</td>\n",
       "      <td>811431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811439</td>\n",
       "      <td>811439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    keep  exact  flip  reverse   both  complement   query  subject\n",
       "2   True  False  True     True  False        True  811358   811358\n",
       "14  True  False  True     True  False        True  811368   811368\n",
       "27  True  False  True     True  False        True  811381   811381\n",
       "41  True  False  True     True  False        True  811395   811395\n",
       "46  True  False  True     True  False        True  811400   811400\n",
       "50  True  False  True     True  False        True  811402   811402\n",
       "65  True  False  True     True  False        True  811415   811415\n",
       "74  True  False  True     True  False        True  811422   811422\n",
       "75  True  False  True     True  False        True  811423   811423\n",
       "83  True  False  True     True  False        True  811431   811431\n",
       "91  True  False  True     True  False        True  811439   811439"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp[tmp.reverse]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "fe62b750-af08-430c-8b5c-dcc3dfd5def2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keep</th>\n",
       "      <th>exact</th>\n",
       "      <th>flip</th>\n",
       "      <th>reverse</th>\n",
       "      <th>both</th>\n",
       "      <th>complement</th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [keep, exact, flip, reverse, both, complement, query, subject]\n",
       "Index: []"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp[tmp.both]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "5e700957-b014-4869-9e03-387b2adb9176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     96\n",
       "False    10\n",
       "Name: flip, dtype: int64"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.flip.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "296146cc-a7ab-41c6-9346-9a450662114f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keep</th>\n",
       "      <th>exact</th>\n",
       "      <th>flip</th>\n",
       "      <th>reverse</th>\n",
       "      <th>both</th>\n",
       "      <th>complement</th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811367</td>\n",
       "      <td>811368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811368</td>\n",
       "      <td>811367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811400</td>\n",
       "      <td>811401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811401</td>\n",
       "      <td>811400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811414</td>\n",
       "      <td>811415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>811415</td>\n",
       "      <td>811414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811418</td>\n",
       "      <td>811419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811419</td>\n",
       "      <td>811418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811440</td>\n",
       "      <td>811441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811441</td>\n",
       "      <td>811440</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     keep  exact   flip  reverse   both  complement   query  subject\n",
       "12  False  False  False    False  False       False  811367   811368\n",
       "13  False  False  False    False  False        True  811368   811367\n",
       "48  False  False  False    False  False        True  811400   811401\n",
       "49  False  False  False    False  False       False  811401   811400\n",
       "65  False  False  False    False  False       False  811414   811415\n",
       "66  False  False  False    False  False        True  811415   811414\n",
       "72  False  False  False    False  False       False  811418   811419\n",
       "73  False  False  False    False  False       False  811419   811418\n",
       "97  False  False  False    False  False       False  811440   811441\n",
       "98  False  False  False    False  False       False  811441   811440"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp[tmp.flip==False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "db968c89-ac63-4dfe-aa86-9e25450b632f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>811366</th>\n",
       "      <td>5</td>\n",
       "      <td>275982</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:275982:C:T</td>\n",
       "      <td>-1.158610</td>\n",
       "      <td>1.623030</td>\n",
       "      <td>0.475317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811367</th>\n",
       "      <td>5</td>\n",
       "      <td>275984</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:275984:A:G</td>\n",
       "      <td>-1.215220</td>\n",
       "      <td>0.968170</td>\n",
       "      <td>0.209418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811368</th>\n",
       "      <td>5</td>\n",
       "      <td>275984</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:275984:A:T</td>\n",
       "      <td>-1.023460</td>\n",
       "      <td>3.179240</td>\n",
       "      <td>0.747513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811369</th>\n",
       "      <td>5</td>\n",
       "      <td>275989</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:275989:T:G</td>\n",
       "      <td>-1.059680</td>\n",
       "      <td>1.854150</td>\n",
       "      <td>0.567647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811370</th>\n",
       "      <td>5</td>\n",
       "      <td>275996</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:275996:C:T</td>\n",
       "      <td>1.908060</td>\n",
       "      <td>1.769520</td>\n",
       "      <td>0.280905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811447</th>\n",
       "      <td>5</td>\n",
       "      <td>306692</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:306692:G:A</td>\n",
       "      <td>0.974559</td>\n",
       "      <td>0.321401</td>\n",
       "      <td>0.005570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811448</th>\n",
       "      <td>5</td>\n",
       "      <td>306695</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:306695:C:T</td>\n",
       "      <td>-1.051520</td>\n",
       "      <td>1.797830</td>\n",
       "      <td>0.558627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811449</th>\n",
       "      <td>5</td>\n",
       "      <td>306703</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:306703:C:T</td>\n",
       "      <td>0.645902</td>\n",
       "      <td>1.310650</td>\n",
       "      <td>0.622147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811450</th>\n",
       "      <td>5</td>\n",
       "      <td>306732</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:306732:C:T</td>\n",
       "      <td>-1.033180</td>\n",
       "      <td>3.113500</td>\n",
       "      <td>0.740011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811451</th>\n",
       "      <td>5</td>\n",
       "      <td>306748</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr5:306748:C:T</td>\n",
       "      <td>1.738350</td>\n",
       "      <td>0.358186</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR     POS REF ALT              SNP      BETA        SE         P\n",
       "811366    5  275982   C   T  chr5:275982:C:T -1.158610  1.623030  0.475317\n",
       "811367    5  275984   A   G  chr5:275984:A:G -1.215220  0.968170  0.209418\n",
       "811368    5  275984   A   T  chr5:275984:A:T -1.023460  3.179240  0.747513\n",
       "811369    5  275989   T   G  chr5:275989:T:G -1.059680  1.854150  0.567647\n",
       "811370    5  275996   C   T  chr5:275996:C:T  1.908060  1.769520  0.280905\n",
       "...     ...     ...  ..  ..              ...       ...       ...       ...\n",
       "811447    5  306692   G   A  chr5:306692:G:A  0.974559  0.321401  0.005570\n",
       "811448    5  306695   C   T  chr5:306695:C:T -1.051520  1.797830  0.558627\n",
       "811449    5  306703   C   T  chr5:306703:C:T  0.645902  1.310650  0.622147\n",
       "811450    5  306732   C   T  chr5:306732:C:T -1.033180  3.113500  0.740011\n",
       "811451    5  306748   C   T  chr5:306748:C:T  1.738350  0.358186  0.000012\n",
       "\n",
       "[86 rows x 8 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "9a90cf13-7321-4e5c-a333-076a3f9b2858",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>keep</th>\n",
       "      <th>exact</th>\n",
       "      <th>flip</th>\n",
       "      <th>reverse</th>\n",
       "      <th>both</th>\n",
       "      <th>complement</th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>811401</td>\n",
       "      <td>811367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>811415</td>\n",
       "      <td>811400</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    keep  exact   flip  reverse   both  complement   query  subject\n",
       "47  True   True  False    False  False       False  811401   811367\n",
       "64  True   True  False    False   True        True  811415   811400"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp[tmp.exact==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "226193ae-01d8-4d49-a7e6-0c7e21cf6772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(tmp['query'] == tmp['subject'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "fa46f434-e82f-430a-b74d-3d5f8f8a2b1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      False\n",
       "1      False\n",
       "2      False\n",
       "3      False\n",
       "4      False\n",
       "       ...  \n",
       "106    False\n",
       "107    False\n",
       "108    False\n",
       "109    False\n",
       "110    False\n",
       "Name: subject, Length: 111, dtype: bool"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.query == tmp.subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "3313f3f8-e5d6-44a9-afec-cf6e2fcdf3f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>811401</th>\n",
       "      <td>5</td>\n",
       "      <td>276395</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:276395:G:A</td>\n",
       "      <td>-1.19077</td>\n",
       "      <td>1.87388</td>\n",
       "      <td>0.525130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811415</th>\n",
       "      <td>5</td>\n",
       "      <td>276938</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:276938:C:G</td>\n",
       "      <td>-1.12324</td>\n",
       "      <td>1.80101</td>\n",
       "      <td>0.532845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR     POS REF ALT              SNP     BETA       SE         P\n",
       "811401    5  276395   G   A  chr5:276395:G:A -1.19077  1.87388  0.525130\n",
       "811415    5  276938   C   G  chr5:276938:C:G -1.12324  1.80101  0.532845"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.loc[[811401,811415],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "7964f7af-6bda-437f-b650-003026e2bbef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>811367</th>\n",
       "      <td>5</td>\n",
       "      <td>275984</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr5:275984:A:G</td>\n",
       "      <td>-1.215220</td>\n",
       "      <td>0.968170</td>\n",
       "      <td>0.209418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811400</th>\n",
       "      <td>5</td>\n",
       "      <td>276395</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr5:276395:G:C</td>\n",
       "      <td>0.852057</td>\n",
       "      <td>0.586238</td>\n",
       "      <td>0.146104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR     POS REF ALT              SNP      BETA        SE         P\n",
       "811367    5  275984   G   A  chr5:275984:A:G -1.215220  0.968170  0.209418\n",
       "811400    5  276395   C   G  chr5:276395:G:C  0.852057  0.586238  0.146104"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa.loc[[811367,811400],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "f892f94f-b6a9-4160-86d6-6c76696c49e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    109\n",
       "True       2\n",
       "Name: exact, dtype: int64"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp.exact.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "34a9c4cd-964b-494c-bddb-0b983e72da72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "306584    2\n",
       "276395    2\n",
       "276938    2\n",
       "275984    2\n",
       "276973    2\n",
       "         ..\n",
       "276216    1\n",
       "276215    1\n",
       "276202    1\n",
       "276191    1\n",
       "306748    1\n",
       "Name: POS, Length: 91, dtype: int64"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.POS.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "d6363549-a1ad-4f47-addd-2e5f3a6bd31b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <td>1</td>\n",
       "      <td>976512</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1:976512:G:A</td>\n",
       "      <td>-1.043870</td>\n",
       "      <td>3.130310</td>\n",
       "      <td>0.738780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6959</th>\n",
       "      <td>1</td>\n",
       "      <td>1333753</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:1333753:G:C</td>\n",
       "      <td>-1.048260</td>\n",
       "      <td>2.806530</td>\n",
       "      <td>0.708771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9458</th>\n",
       "      <td>1</td>\n",
       "      <td>1495680</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:1495680:T:C</td>\n",
       "      <td>0.147934</td>\n",
       "      <td>0.850903</td>\n",
       "      <td>0.861979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11995</th>\n",
       "      <td>1</td>\n",
       "      <td>1737811</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:1737811:T:C</td>\n",
       "      <td>-0.046021</td>\n",
       "      <td>1.078830</td>\n",
       "      <td>0.965974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13127</th>\n",
       "      <td>1</td>\n",
       "      <td>1955424</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:1955424:T:C</td>\n",
       "      <td>0.140340</td>\n",
       "      <td>0.104258</td>\n",
       "      <td>0.178278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2991282</th>\n",
       "      <td>22</td>\n",
       "      <td>37206803</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:37206803:C:T</td>\n",
       "      <td>2.635060</td>\n",
       "      <td>1.949880</td>\n",
       "      <td>0.176570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2999438</th>\n",
       "      <td>22</td>\n",
       "      <td>39240918</td>\n",
       "      <td>T</td>\n",
       "      <td>TGCGC</td>\n",
       "      <td>chr22:39240918:T:TGCGC</td>\n",
       "      <td>1.020610</td>\n",
       "      <td>1.452690</td>\n",
       "      <td>0.482324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3013185</th>\n",
       "      <td>22</td>\n",
       "      <td>44786566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr22:44786566:G:A</td>\n",
       "      <td>0.580112</td>\n",
       "      <td>0.696360</td>\n",
       "      <td>0.404809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3024911</th>\n",
       "      <td>22</td>\n",
       "      <td>50315712</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr22:50315712:A:G</td>\n",
       "      <td>-1.085230</td>\n",
       "      <td>2.524020</td>\n",
       "      <td>0.667223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3028072</th>\n",
       "      <td>22</td>\n",
       "      <td>50603663</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr22:50603663:G:A</td>\n",
       "      <td>-1.089630</td>\n",
       "      <td>2.563200</td>\n",
       "      <td>0.670760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR       POS REF    ALT                     SNP      BETA        SE  \\\n",
       "1733       1    976512   G      A         chr1:976512:G:A -1.043870  3.130310   \n",
       "6959       1   1333753   G      C        chr1:1333753:G:C -1.048260  2.806530   \n",
       "9458       1   1495680   T      C        chr1:1495680:T:C  0.147934  0.850903   \n",
       "11995      1   1737811   T      C        chr1:1737811:T:C -0.046021  1.078830   \n",
       "13127      1   1955424   T      C        chr1:1955424:T:C  0.140340  0.104258   \n",
       "...      ...       ...  ..    ...                     ...       ...       ...   \n",
       "2991282   22  37206803   C      T      chr22:37206803:C:T  2.635060  1.949880   \n",
       "2999438   22  39240918   T  TGCGC  chr22:39240918:T:TGCGC  1.020610  1.452690   \n",
       "3013185   22  44786566   G      A      chr22:44786566:G:A  0.580112  0.696360   \n",
       "3024911   22  50315712   A      G      chr22:50315712:A:G -1.085230  2.524020   \n",
       "3028072   22  50603663   G      A      chr22:50603663:G:A -1.089630  2.563200   \n",
       "\n",
       "                P  \n",
       "1733     0.738780  \n",
       "6959     0.708771  \n",
       "9458     0.861979  \n",
       "11995    0.965974  \n",
       "13127    0.178278  \n",
       "...           ...  \n",
       "2991282  0.176570  \n",
       "2999438  0.482324  \n",
       "3013185  0.404809  \n",
       "3024911  0.667223  \n",
       "3028072  0.670760  \n",
       "\n",
       "[1000 rows x 8 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "c640660f-a640-4e09-ab75-640509524751",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pandas(Index=1733, CHR=1, POS=976512, REF='G', ALT='A', SNP='chr1:976512:G:A', BETA=-1.04387, SE=3.13031, P=0.7387797791025819) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=6959, CHR=1, POS=1333753, REF='G', ALT='C', SNP='chr1:1333753:G:C', BETA=-1.04826, SE=2.80653, P=0.7087710984181352) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=9458, CHR=1, POS=1495680, REF='T', ALT='C', SNP='chr1:1495680:T:C', BETA=0.147934, SE=0.850903, P=0.861979425862772) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=11995, CHR=1, POS=1737811, REF='T', ALT='C', SNP='chr1:1737811:T:C', BETA=-0.0460207, SE=1.07883, P=0.9659741397427256) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=13127, CHR=1, POS=1955424, REF='T', ALT='C', SNP='chr1:1955424:T:C', BETA=0.14034, SE=0.104258, P=0.178277690755054) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=14530, CHR=1, POS=2303171, REF='C', ALT='T', SNP='chr1:2303171:C:T', BETA=-1.05115, SE=2.34098, P=0.6534178585573059) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=15435, CHR=1, POS=2406725, REF='G', ALT='A', SNP='chr1:2406725:G:A', BETA=-1.11988, SE=1.3376, P=0.4024602568261106) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=31462, CHR=1, POS=10307349, REF='G', ALT='A', SNP='chr1:10307349:G:A', BETA=-1.10213, SE=1.49971, P=0.4624001858860571) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=33819, CHR=1, POS=11080678, REF='T', ALT='C', SNP='chr1:11080678:T:C', BETA=-1.65391, SE=1.53434, P=0.2810657974090764) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n",
      "Pandas(Index=38760, CHR=1, POS=12348810, REF='C', ALT='T', SNP='chr1:12348810:C:T', BETA=2.45887, SE=1.90922, P=0.1977839290228308) Pandas(Index=301220, CHR=1, POS=242107640, REF='T', ALT='C', SNP='chr1:242107640:T:C', BETA=0.331632, SE=0.441286, P=0.452343131230923)\n"
     ]
    }
   ],
   "source": [
    "smry = []\n",
    "query = a[:10].itertuples()\n",
    "subject = a[100:210].itertuples()\n",
    "qi,si = next(query,None),next(subject,None)\n",
    "multi_snps = []\n",
    "while(qi and si):\n",
    "    if qi[1]>si[1]:\n",
    "        si = next(subject,None)\n",
    "        multi_snps = []\n",
    "        continue\n",
    "    elif qi[1]<si[1]:\n",
    "        qi = next(query,None)\n",
    "        if len(multi_snps)==0:\n",
    "            smry.append([False]*5+[-1,-1])\n",
    "        else:\n",
    "            for s in multi_snps:\n",
    "                smry.append(snp_match(qi[3],qi[4],s[3],s[4])+[qi[0],s[0]])\n",
    "        continue\n",
    "    else:\n",
    "        if qi[2]>si[2]:\n",
    "            si = next(subject,None)\n",
    "            multi_snps = []\n",
    "            continue\n",
    "        elif qi[2]<si[2]:\n",
    "            qi = next(query,None)\n",
    "            if len(multi_snps)==0:\n",
    "                smry.append(np.array([False]*5))\n",
    "            else:\n",
    "                for s in multi_snps:\n",
    "                    smry.append(snp_match(qi[3],qi[4],s[3],s[4])+[qi[0],s[0]])\n",
    "            continue\n",
    "        else:\n",
    "            #same pos has multiple snps\n",
    "            #query compare with each of them in subject\n",
    "            multi_snps.append(si)\n",
    "            smry.append(snp_match(qi[3],qi[4],si[3],si[4])+[qi[0],si[0]])\n",
    "            si = next(subject,None)\n",
    "smry = pd.DataFrame(smry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "bc2ea67a-c24d-4701-b262-982a8b7c8014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False]),\n",
       " array([False, False, False, False, False])]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1a47e18d-8433-4ac1-9850-a9c78284b08e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ai = a[:5].itertuples()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2a4d0a41-90e9-46b7-87c8-42f28a3cb3b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pandas(Index=1733, CHR=1, POS=976512, REF='G', ALT='A', SNP='chr1:976512:G:A', BETA=-1.04387, SE=3.13031, P=0.7387797791025819)\n",
      "tmp Pandas(Index=6959, CHR=1, POS=1333753, REF='G', ALT='C', SNP='chr1:1333753:G:C', BETA=-1.04826, SE=2.80653, P=0.7087710984181352)\n",
      "Pandas(Index=9458, CHR=1, POS=1495680, REF='T', ALT='C', SNP='chr1:1495680:T:C', BETA=0.147934, SE=0.850903, P=0.861979425862772)\n",
      "tmp Pandas(Index=11995, CHR=1, POS=1737811, REF='T', ALT='C', SNP='chr1:1737811:T:C', BETA=-0.0460207, SE=1.07883, P=0.9659741397427256)\n",
      "Pandas(Index=13127, CHR=1, POS=1955424, REF='T', ALT='C', SNP='chr1:1955424:T:C', BETA=0.14034, SE=0.104258, P=0.178277690755054)\n",
      "tmp None\n"
     ]
    }
   ],
   "source": [
    "for i in ai:\n",
    "    print(i)\n",
    "    tmp = next(ai,None)\n",
    "    print('tmp',tmp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc5b4ae8-26a0-4071-9995-d76bc8128423",
   "metadata": {},
   "source": [
    "### 3.SNPs match in two sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "a762ff63-56d6-4b71-8ac7-d272988af753",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snps_match(query,subject,keep_ambiguous=True):\n",
    "    query.index = query.iloc[:,:2].astype(str).agg(':'.join, axis=1)\n",
    "    subject.index = subject.iloc[:,:2].astype(str).agg(':'.join, axis=1)\n",
    "    #overlap snps by chr+pos\n",
    "    print(\"Total rows of query: \",query.shape[0],\"Total rows of subject: \",subject.shape[0])\n",
    "    subject = subject[subject.index.isin(query.index)]\n",
    "    query = query.loc[subject.index]\n",
    "    print(\"Overlap chr:pos\",query.shape[0])\n",
    "    if query.index.duplicated().any():\n",
    "        raise Exception(\"There are duplicated chr:pos\")\n",
    "    pm = pair_match(query.ALT,query.REF,subject.ALT,subject.REF)\n",
    "    if keep_ambiguous:\n",
    "        print('Warning: there are',sum(~pm.ambiguous),'ambiguous SNPs')\n",
    "        pm = pm.iloc[:,1:]\n",
    "    else:\n",
    "        pm = pm[~pm.ambiguous].iloc[:,1:]\n",
    "    keep_idx = pm.any(axis=1)\n",
    "    print(\"Overlap SNPs\",sum(keep_idx))\n",
    "    #overlap snps by chr+pos+alleles.\n",
    "    new_subject = subject[keep_idx]\n",
    "    #update beta and snp info\n",
    "    new_query = pd.concat([new_subject.iloc[:,:5],query[keep_idx].iloc[:,5:]],axis=1)\n",
    "    new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n",
    "    return new_query,new_subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "id": "bf75a189-62db-421b-9000-e97ae8c0b14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pair_match(a1,a2,ref1,ref2):\n",
    "    # a1 and a2 are the first data-set\n",
    "\t# ref1 and ref2 are the 2nd data-set\n",
    "\t# Make all the alleles into upper-case, as A,T,C,G:\n",
    "    a1 = a1.str.upper()\n",
    "    a2 = a2.str.upper()\n",
    "    ref1 = ref1.str.upper()\n",
    "    ref2 = ref2.str.upper()\n",
    "\t# Strand flip, to change the allele representation in the 2nd data-set\n",
    "    flip1 = ref1.apply(strand_flip)\n",
    "    flip2 = ref2.apply(strand_flip)\n",
    "    result = {}\n",
    "    result[\"ambiguous\"] = ((a1==\"A\") & (a2==\"T\")) | ((a1==\"T\") & (a2==\"A\")) | ((a1==\"C\") & (a2==\"G\")) | ((a1==\"G\") & (a2==\"C\"))\n",
    "    # as long as scenario 1 is involved, sign_flip will return TRUE\n",
    "    result[\"sign_flip\"] = ((a1==ref2) & (a2==ref1)) | ((a1==flip2) & (a2==flip1))\n",
    "\t# as long as scenario 2 is involved, strand_flip will return TRUE\n",
    "    result[\"strand_flip\"] = ((a1==flip1) & (a2==flip2)) | ((a1==flip2) & (a2==flip1))\n",
    "\t# remove other cases, eg, tri-allelic, one dataset is A C, the other is A G, for example.\n",
    "    result[\"exact_match\"] = ((a1 == ref1) & (a2 == ref2))\n",
    "    return pd.DataFrame(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "id": "25b9e18d-8cd2-49fb-97f5-1c93713248b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def strand_flip(s):\n",
    "    return ''.join(Seq(s).reverse_complement())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ef6957-b7a0-468b-8293-17242c2f87d1",
   "metadata": {},
   "source": [
    "### 4.Create testing MWE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "14b50428-9943-47d1-9d08-283246f435a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "exome_sumstats = Sumstat(sumstats_path)\n",
    "#exome_sumstats.extractbyregion(region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "4b640270-d379-4c13-afc4-cc7e5cafa77f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "a=exome_sumstats.ss.sample(n=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "id": "724f822a-e7e8-4bfa-a2f1-977ef95335a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = a.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "id": "db80bb59-0e14-41f2-bc49-751e56a50977",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>1</td>\n",
       "      <td>953305</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1:953305:G:A</td>\n",
       "      <td>-0.137062</td>\n",
       "      <td>0.731144</td>\n",
       "      <td>0.851298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118</th>\n",
       "      <td>1</td>\n",
       "      <td>1014359</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:1014359:G:T</td>\n",
       "      <td>-1.110110</td>\n",
       "      <td>1.724070</td>\n",
       "      <td>0.519648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4067</th>\n",
       "      <td>1</td>\n",
       "      <td>1211543</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:1211543:G:T</td>\n",
       "      <td>-0.220586</td>\n",
       "      <td>0.947774</td>\n",
       "      <td>0.815963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7641</th>\n",
       "      <td>1</td>\n",
       "      <td>1355779</td>\n",
       "      <td>GA</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:1355779:GA:G</td>\n",
       "      <td>0.096933</td>\n",
       "      <td>0.488808</td>\n",
       "      <td>0.842807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13114</th>\n",
       "      <td>1</td>\n",
       "      <td>1955369</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:1955369:C:G</td>\n",
       "      <td>-0.066739</td>\n",
       "      <td>0.762243</td>\n",
       "      <td>0.930230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3022270</th>\n",
       "      <td>22</td>\n",
       "      <td>50210362</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:50210362:C:T</td>\n",
       "      <td>-1.147250</td>\n",
       "      <td>1.521600</td>\n",
       "      <td>0.450867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3023900</th>\n",
       "      <td>22</td>\n",
       "      <td>50278367</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:50278367:C:T</td>\n",
       "      <td>-1.222950</td>\n",
       "      <td>1.248900</td>\n",
       "      <td>0.327472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3025632</th>\n",
       "      <td>22</td>\n",
       "      <td>50455242</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr22:50455242:G:A</td>\n",
       "      <td>0.634902</td>\n",
       "      <td>0.433051</td>\n",
       "      <td>0.142618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3026405</th>\n",
       "      <td>22</td>\n",
       "      <td>50488346</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr22:50488346:A:G</td>\n",
       "      <td>-0.564272</td>\n",
       "      <td>0.900978</td>\n",
       "      <td>0.531125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3026963</th>\n",
       "      <td>22</td>\n",
       "      <td>50517779</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr22:50517779:C:T</td>\n",
       "      <td>0.989030</td>\n",
       "      <td>0.882422</td>\n",
       "      <td>0.262367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR       POS REF ALT                 SNP      BETA        SE  \\\n",
       "709        1    953305   G   A     chr1:953305:G:A -0.137062  0.731144   \n",
       "2118       1   1014359   G   T    chr1:1014359:G:T -1.110110  1.724070   \n",
       "4067       1   1211543   G   T    chr1:1211543:G:T -0.220586  0.947774   \n",
       "7641       1   1355779  GA   G   chr1:1355779:GA:G  0.096933  0.488808   \n",
       "13114      1   1955369   C   G    chr1:1955369:C:G -0.066739  0.762243   \n",
       "...      ...       ...  ..  ..                 ...       ...       ...   \n",
       "3022270   22  50210362   C   T  chr22:50210362:C:T -1.147250  1.521600   \n",
       "3023900   22  50278367   C   T  chr22:50278367:C:T -1.222950  1.248900   \n",
       "3025632   22  50455242   G   A  chr22:50455242:G:A  0.634902  0.433051   \n",
       "3026405   22  50488346   A   G  chr22:50488346:A:G -0.564272  0.900978   \n",
       "3026963   22  50517779   C   T  chr22:50517779:C:T  0.989030  0.882422   \n",
       "\n",
       "                P  \n",
       "709      0.851298  \n",
       "2118     0.519648  \n",
       "4067     0.815963  \n",
       "7641     0.842807  \n",
       "13114    0.930230  \n",
       "...           ...  \n",
       "3022270  0.450867  \n",
       "3023900  0.327472  \n",
       "3025632  0.142618  \n",
       "3026405  0.531125  \n",
       "3026963  0.262367  \n",
       "\n",
       "[1000 rows x 8 columns]"
      ]
     },
     "execution_count": 488,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "id": "499e9bce-6a25-4e71-8737-d5612baec9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss1 = a[20:520].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "a0c6489c-50b5-4b48-9c2f-773b599550bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss1 = ss1[~ss1.POS.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "80150add-0303-4c7a-9774-f3c06d71d37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_refalt(ss):\n",
    "    ss = ss.copy()\n",
    "    ref = ss.REF.copy()\n",
    "    ss.REF = ss.ALT\n",
    "    ss.ALT = ref\n",
    "    ss.BETA = -ss.BETA\n",
    "    return ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "id": "0f96309e-d6ed-4b12-b275-684a2302ddb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flip_snps(ss):\n",
    "    ss = ss.copy()\n",
    "    ss.REF = [strand_flip(i) for i in ss.REF]\n",
    "    ss.ALT = [strand_flip(i) for i in ss.ALT]\n",
    "    return ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "id": "027efd41-761d-4b99-97b4-0a00c42d1b78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total rows of query:  500 Total rows of subject:  1000\n",
      "Overlap chr:pos 500\n",
      "Warning: there are 418 ambiguous SNPs\n",
      "Overlap SNPs 500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/1988675.1.plot.q/ipykernel_30609/2132708780.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(              CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850   1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns],\n",
       "               CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850  -1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns])"
      ]
     },
     "execution_count": 498,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_match(flip_snps(ss1),a,keep_ambiguous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "eeabf2be-7243-4040-846e-c65dc8b6b970",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total rows of query:  500 Total rows of subject:  1000\n",
      "Overlap chr:pos 500\n",
      "Warning: there are 418 ambiguous SNPs\n",
      "Overlap SNPs 500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/1988675.1.plot.q/ipykernel_30609/2132708780.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(              CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850  -1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns],\n",
       "               CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850  -1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns])"
      ]
     },
     "execution_count": 499,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_match(reverse_refalt(ss1),a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "fdf9b14e-d4aa-44c2-a155-f91f3484603a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total rows of query:  500 Total rows of subject:  1000\n",
      "Overlap chr:pos 500\n",
      "Warning: there are 418 ambiguous SNPs\n",
      "Overlap SNPs 500\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/1988675.1.plot.q/ipykernel_30609/2132708780.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  new_query.BETA[pm.sign_flip] = -new_query.BETA[pm.sign_flip]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(              CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850  -1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns],\n",
       "               CHR        POS         REF ALT                         SNP  \\\n",
       " 1:18744859      1   18744859           G   A           chr1:18744859:G:A   \n",
       " 1:19112524      1   19112524           C   T           chr1:19112524:C:T   \n",
       " 1:19112744      1   19112744           C   T           chr1:19112744:C:T   \n",
       " 1:19220870      1   19220870  TTCACACCGA   T  chr1:19220870:TTCACACCGA:T   \n",
       " 1:19324561      1   19324561           G   A           chr1:19324561:G:A   \n",
       " ...           ...        ...         ...  ..                         ...   \n",
       " 10:93074445    10   93074445           G   A          chr10:93074445:G:A   \n",
       " 10:94171346    10   94171346           C   T          chr10:94171346:C:T   \n",
       " 10:97370850    10   97370850           A   T          chr10:97370850:A:T   \n",
       " 10:100293231   10  100293231           A   G         chr10:100293231:A:G   \n",
       " 10:101803994   10  101803994           A   G         chr10:101803994:A:G   \n",
       " \n",
       "                   BETA        SE         P  \n",
       " 1:18744859   -1.280560  1.327880  0.334864  \n",
       " 1:19112524   -1.063210  2.760900  0.700168  \n",
       " 1:19112744   -0.475555  0.539997  0.378501  \n",
       " 1:19220870   -1.101250  2.434530  0.651020  \n",
       " 1:19324561    0.908711  1.011330  0.368904  \n",
       " ...                ...       ...       ...  \n",
       " 10:93074445  -1.282030  0.989365  0.195040  \n",
       " 10:94171346  -1.054960  2.390500  0.658984  \n",
       " 10:97370850  -1.069910  2.301690  0.642049  \n",
       " 10:100293231  0.511226  1.280380  0.689689  \n",
       " 10:101803994 -1.083250  2.012430  0.590386  \n",
       " \n",
       " [500 rows x 8 columns])"
      ]
     },
     "execution_count": 500,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_match(flip_snps(reverse_refalt(ss1)),a)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad6d7462-e221-4a13-8edf-4510b4a2f593",
   "metadata": {},
   "source": [
    "### 4.1 Write out test examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "id": "0663a066-09a3-4328-bd05-3ff7df4fadbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "a.to_csv('data/testflip/snps1000.regenie.snp_stats.gz', sep = \"\\t\", header = True, index = False,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "id": "26deafe9-3a72-4383-a449-0b22836980ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "ss1.columns = ['CHR','POS','REF','ALT','SNP','BETA','SE','P']\n",
    "ss1.to_csv('data/testflip/snps500.regenie.snp_stats.gz', sep = \"\\t\", header = True, index = False,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "fbba9498-29eb-4d83-8869-2a5aadf4f721",
   "metadata": {},
   "outputs": [],
   "source": [
    "fss1 = flip_snps(ss1)\n",
    "fss1.columns = ['CHR','POS','A0','A1','SNP','STAT','SE','P']\n",
    "fss1.to_csv('data/testflip/flip/snps500_flip.regenie.snp_stats.gz', sep = \"\\t\", header = True, index = False,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "id": "0bfdba54-877a-46c4-ad60-34d1caeae5b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reverse_refalt(ss1).to_csv('data/testflip/snps500_rea0a1.regenie.snp_stats.gz', sep = \"\\t\", header = True, index = False,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "id": "6f018a8f-e7cf-4ba0-b8d8-77d4989ee301",
   "metadata": {},
   "outputs": [],
   "source": [
    "flip_snps(reverse_refalt(ss1)).to_csv('data/testflip/snps500_flip_rea0a1.regenie.snp_stats.gz', sep = \"\\t\", header = True, index = False,compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb7eb7a-c88a-44db-92b2-ba2cae05d93e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
